// @postman/runtime/dist/items/mcp/utils/ClientWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientWrapper = void 0;
class ClientWrapper {
    client;
    pendingRequests;
    abortController;
    constructor({ client, pendingRequests, abortController, }) {
        this.client = client;
        this.pendingRequests =
            pendingRequests ??
                new Map();
        this.abortController = abortController;
    }
    getClient() {
        return this.client;
    }
    getPendingRequests() {
        return this.pendingRequests;
    }
    getAbortController() {
        return this.abortController;
    }
    resetAbortController(event, incoming) {
        const abortListener = () => {
            incoming.enqueue({
                type: 'mcp:aborted',
                payload: {
                    message: `Operation was cancelled: ${event.type}`,
                },
                timestamp: new Date().toISOString(),
            });
        };
        this.abortController = new AbortController();
        this.abortController.signal.addEventListener('abort', abortListener);
        return this.abortController;
    }
}
exports.ClientWrapper = ClientWrapper;
