// @postman/runtime.grpc-utils/dist/generate-json-schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = generateJSONSchema;
const protobufjs_1 = __importDefault(require("@postman/protobufjs"));
const js_base64_1 = require("js-base64");
// Generates a JSON Schema that describes the given Protobuf.Type.
function generateJSONSchema(rootType) {
    const definitions = {};
    const jsonPointers = new Map();
    const requiredByUser = new Set();
    const mapFieldPaths = [];
    const currFieldPath = [];
    return {
        $schema: 'http://json-schema.org/draft-07/schema',
        definitions,
        ...declareSchema(rootType),
        // caching the paths to map fields so that we can
        // efficiently sanitize the payload at runtime
        mapFields: mapFieldPaths,
    };
    // Returns a (memoized) subschema that describes the given Type or Enum.
    function declareSchema(type) {
        let jsonPointer = jsonPointers.get(type);
        if (!jsonPointer) {
            jsonPointer = `#/definitions/${type.fullName}`;
            jsonPointers.set(type, jsonPointer);
            definitions[type.fullName] = generateSchema(type);
        }
        return { $ref: jsonPointer };
    }
    // Generates a new (not memoized) subschema from the given Type or Enum.
    function generateSchema(type) {
        if (type instanceof protobufjs_1.default.Enum) {
            if (type.fullName === '.google.protobuf.NullValue') {
                return {
                    title: 'google.protobuf.NullValue',
                    description: 'A NullValue can be represented as null',
                    enum: [null, 'NULL_VALUE'],
                    examples: [null],
                };
            }
            const enumValues = [
                ...Object.keys(type.values), // Enum names (strings)
                ...Object.values(type.values), // Enum ids (integers)
                null,
            ];
            return {
                enum: Array.from(new Set(enumValues)),
                title: `enum ${type.name}`,
                description: `${Object.entries(type.values)
                    .map(([k, v]) => `${JSON.stringify(k)} (${v})`)
                    .join(' | ')}`,
                default: type.valuesById[0] || 0,
            };
        }
        if (type instanceof protobufjs_1.default.Type) {
            // Special handling for google.protobuf.Value
            if (type.fullName === '.google.protobuf.Value') {
                return {
                    anyOf: [
                        { type: 'null' },
                        { type: 'string' },
                        { type: 'number' },
                        { type: 'boolean' },
                        {
                            type: 'array',
                            items: true, // Allow any type of items
                        },
                        {
                            type: 'object',
                            additionalProperties: true, // Allow any properties
                        },
                    ],
                    title: 'google.protobuf.Value',
                    description: 'A dynamically typed value that can be null, string, number, boolean, list, or struct',
                    default: null,
                };
            }
            // Special handling for google.protobuf.Struct
            if (type.fullName === '.google.protobuf.Struct') {
                return {
                    type: ['object', 'null'],
                    additionalProperties: true,
                    title: 'google.protobuf.Struct',
                    description: 'A JSON object with arbitrary key-value pairs',
                    default: {},
                    examples: [
                        {
                            foo: 'bar',
                        },
                    ],
                };
            }
            // Handle well-known Google protobuf types specially
            if (type.fullName === '.google.protobuf.Duration' ||
                type.fullName === 'Duration') {
                return {
                    title: 'google.protobuf.Duration',
                    description: 'A Duration can be represented as a string (e.g., "1s", "1m", "1h", "1.5s", "-1s", "1m10s", "2h30m", "100ms", "500us", "1000ns") or as an object with seconds and nanos',
                    anyOf: [
                        {
                            type: 'object',
                            properties: {
                                seconds: {
                                    type: ['integer', 'string'],
                                    description: 'Signed seconds of the span',
                                },
                                nanos: {
                                    type: 'integer',
                                    description: 'Signed fractions of a second at nanosecond resolution',
                                },
                            },
                            required: ['seconds'],
                        },
                        {
                            type: 'string',
                            pattern: '^-?(\\d+(\\.\\d+)?(s|m|h|ms|us|μs|ns))+$',
                            description: 'Duration string in format like "1s", "1m", "1h", "1.5s", "-1s", "1m10s", "2h30m", "100ms", "500us", "1000ns"',
                        },
                    ],
                    examples: [
                        '1s',
                        '1m',
                        '1h',
                        '1.5s',
                        '-1s',
                        '1m10s',
                        '2h30m',
                        '100ms',
                        '500us',
                        '1000ns',
                    ],
                };
            }
            if (type.fullName === '.google.protobuf.Timestamp' ||
                type.fullName === 'Timestamp') {
                return {
                    title: 'google.protobuf.Timestamp',
                    description: 'A Timestamp can be represented as an ISO 8601 string or as an object with seconds and nanos',
                    anyOf: [
                        {
                            type: 'object',
                            properties: {
                                seconds: {
                                    type: ['integer', 'string'],
                                    description: 'Represents seconds of UTC time since Unix epoch',
                                },
                                nanos: {
                                    type: 'integer',
                                    description: 'Non-negative fractions of a second at nanosecond resolution',
                                },
                            },
                            required: ['seconds'],
                        },
                        {
                            type: 'string',
                            pattern: '^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d{1,3})?(Z|[+-]\\d{2}:\\d{2})?$',
                            description: 'ISO 8601 timestamp string (e.g., "2024-01-15T10:30:00Z", "2024-01-15T10:30:00+05:30")',
                        },
                    ],
                    examples: [new Date().toISOString()],
                };
            }
            if (type.fullName === '.google.protobuf.ListValue') {
                return {
                    title: 'google.protobuf.ListValue',
                    description: 'A ListValue can be represented as an array',
                    anyOf: [
                        {
                            type: 'array',
                            items: true,
                        },
                        {
                            type: 'object',
                            properties: {
                                values: {
                                    type: 'array',
                                    items: true,
                                },
                            },
                            required: ['values'],
                        },
                    ],
                    examples: [
                        ['hello', 'world'],
                        [1, 2, 3],
                    ],
                };
            }
            // Handle wrapper types (BoolValue, DoubleValue, etc.)
            if (type.fullName === '.google.protobuf.BoolValue') {
                return {
                    title: 'google.protobuf.BoolValue',
                    description: 'A BoolValue can be represented as a boolean.',
                    anyOf: [
                        {
                            type: ['boolean', 'null'],
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: ['boolean', 'null'],
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: [true, false],
                };
            }
            if (type.fullName === '.google.protobuf.DoubleValue') {
                return {
                    title: 'google.protobuf.DoubleValue',
                    description: 'A DoubleValue can be represented as a number.',
                    anyOf: [
                        {
                            type: ['number', 'null'],
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: ['number', 'null'],
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: [3.14159],
                };
            }
            if (type.fullName === '.google.protobuf.FloatValue') {
                return {
                    title: 'google.protobuf.FloatValue',
                    description: 'A FloatValue can be represented as a number.',
                    anyOf: [
                        {
                            type: ['number', 'null'],
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: ['number', 'null'],
                                },
                            },
                        },
                    ],
                    examples: [3.14],
                };
            }
            if (type.fullName === '.google.protobuf.Int32Value') {
                return {
                    title: 'google.protobuf.Int32Value',
                    description: 'An Int32Value can be represented as an integer',
                    anyOf: [
                        {
                            type: 'integer',
                            minimum: -0x80000000,
                            maximum: 0x7fffffff,
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: 'integer',
                                    minimum: -0x80000000,
                                    maximum: 0x7fffffff,
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: [42, -100],
                };
            }
            if (type.fullName === '.google.protobuf.UInt32Value') {
                return {
                    title: 'google.protobuf.UInt32Value',
                    description: 'A UInt32Value can be represented as an integer (unsigned)',
                    anyOf: [
                        {
                            type: 'integer',
                            minimum: 0,
                            maximum: 0xffffffff,
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: 'integer',
                                    minimum: 0,
                                    maximum: 0xffffffff,
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: [100, 4294967295],
                };
            }
            if (type.fullName === '.google.protobuf.Int64Value') {
                return {
                    title: 'google.protobuf.Int64Value',
                    description: 'An Int64Value can be represented as an integer/string',
                    anyOf: [
                        {
                            type: ['integer', 'string'],
                            pattern: '^-?[0-9]+$',
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: ['integer', 'string'],
                                    pattern: '^-?[0-9]+$',
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: [42, '-9223372036854'],
                };
            }
            if (type.fullName === '.google.protobuf.UInt64Value') {
                return {
                    title: 'google.protobuf.UInt64Value',
                    description: 'A UInt64Value can be represented as an integer/string (unsigned)',
                    anyOf: [
                        {
                            type: ['integer', 'string'],
                            pattern: '^[0-9]+$',
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: ['integer', 'string'],
                                    pattern: '^[0-9]+$',
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: ['18446744073709551615'],
                };
            }
            if (type.fullName === '.google.protobuf.BytesValue') {
                return {
                    title: 'google.protobuf.BytesValue',
                    description: 'A BytesValue can be represented as a string (base64)',
                    anyOf: [
                        {
                            type: 'string',
                            pattern: '^([a-zA-Z0-9+/]{4})*([a-zA-Z0-9+/]{2}(==)?|[a-zA-Z0-9+/]{3}=?)?$',
                        },
                        {
                            type: 'object',
                            properties: {
                                value: {
                                    type: 'string',
                                    pattern: '^([a-zA-Z0-9+/]{4})*([a-zA-Z0-9+/]{2}(==)?|[a-zA-Z0-9+/]{3}=?)?$',
                                },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: ['SGVsbG8gV29ybGQ='],
                };
            }
            if (type.fullName === '.google.protobuf.StringValue') {
                return {
                    title: 'google.protobuf.StringValue',
                    description: 'A StringValue can be represented as a string',
                    anyOf: [
                        { type: 'string' },
                        {
                            type: 'object',
                            properties: {
                                value: { type: 'string' },
                            },
                            required: ['value'],
                        },
                    ],
                    examples: ['hello world', ''],
                };
            }
            if (type.fullName === '.google.protobuf.Any') {
                return {
                    title: 'google.protobuf.Any',
                    type: ['object', 'null'],
                    description: 'Any contains an arbitrary serialized message along with a URL that describes the type of the serialized message. Can be represented as an object with @type and the message properties, or as an object with type_url and value fields.',
                    anyOf: [
                        {
                            type: 'object',
                            properties: {
                                '@type': {
                                    type: 'string',
                                },
                            },
                            required: ['@type'],
                            additionalProperties: true,
                        },
                        {
                            type: 'object',
                            properties: {
                                type_url: {
                                    type: 'string',
                                },
                                value: {
                                    type: 'string',
                                },
                            },
                            required: ['type_url', 'value'],
                            additionalProperties: false,
                        },
                    ],
                    examples: [
                        {
                            '@type': 'type.googleapis.com/google.protobuf.StringValue',
                            value: 'hello world',
                        },
                        {
                            '@type': 'type.googleapis.com/google.protobuf.BoolValue',
                            value: true,
                        },
                    ],
                };
            }
            let isEmpty = true;
            const properties = {};
            const required = [];
            for (const field of Object.values(type.fields)) {
                isEmpty = false;
                currFieldPath.push(field.name);
                applyField(field, properties, required);
                currFieldPath.pop();
            }
            if (!isEmpty) {
                // Only non-empty message types are actually required. Everything else has a default value.
                requiredByUser.add(type);
            }
            return {
                type: isEmpty ? ['object', 'null'] : 'object',
                properties,
                required,
                additionalProperties: { title: 'Unknown field' },
                title: `message ${type.name}`,
                description: `{ ${Object.keys(type.fields).join(', ')} }`,
                default: {},
            };
        }
        console.warn(`Unrecognized Protobuf class: ${type && type.constructor.name}`);
        return {};
    }
    // Augments a "properties" object and "required" array with a given field.
    function applyField(field, properties, required) {
        let schema;
        if (field.resolvedType) {
            schema = declareSchema(field.resolvedType);
        }
        else {
            schema = getKnownSchema(field.type);
            schema.title = field.type;
            schema.default = normalizeProtobufValue(field.typeDefault);
        }
        if (field.map) {
            const keyType = field.keyType;
            schema = {
                oneOf: [
                    // JSON representation
                    {
                        type: ['object', 'null'],
                        additionalProperties: schema,
                        title: `map<${keyType}, ${field.type}>`,
                        default: {},
                    },
                    // Array representation for backward compatibility
                    // Due to how we previously decoded map fields, we need to support this
                    {
                        type: ['array', 'null'],
                        items: {
                            type: 'object',
                            properties: {
                                key: (() => {
                                    const keySchema = getKnownSchema(keyType);
                                    keySchema.title = keyType;
                                    return keySchema;
                                })(),
                                value: schema,
                            },
                            required: ['key', 'value'],
                            additionalProperties: false,
                        },
                        title: `repeated { "key: ${keyType}, "value": ${field.type} }`,
                        default: [],
                    },
                ],
                title: `map<${keyType}, ${field.type}>`,
                default: {},
            };
            mapFieldPaths.push(currFieldPath.join('.'));
        }
        else if (field.repeated) {
            schema = {
                type: ['array', 'null'],
                items: schema,
                title: `repeated ${field.type}`,
                default: [],
            };
        }
        else if (field.required) {
            if (requiredByUser.has(field.resolvedType)) {
                required.push(field.name);
            }
        }
        properties[field.name] = schema;
    }
}
// Returns the subschema of a well-known Protobuf type.
function getKnownSchema(typeName) {
    switch (typeName) {
        case 'bool':
            return {
                type: ['boolean', 'null'],
            };
        case 'string':
            return {
                type: ['string', 'null'],
            };
        case 'bytes':
            return {
                type: ['string', 'array', 'null'],
                pattern: '^([a-zA-Z0-9+/]{4})*([a-zA-Z0-9+/]{2}(==)?|[a-zA-Z0-9+/]{3}=?)?$',
                items: {
                    type: 'integer',
                    default: 0,
                    minimum: 0,
                    maximum: 255,
                },
                description: 'A string of base64, or an array of numbers between 0 and 255',
            };
        case 'int32':
        case 'sint32':
        case 'sfixed32':
            return {
                type: ['integer', 'null'],
                minimum: -0x80000000,
                maximum: 0x7fffffff,
                description: 'An integer between -2147483648 and 2147483647',
            };
        case 'uint32':
        case 'fixed32':
            return {
                type: ['integer', 'null'],
                minimum: 0,
                maximum: 0xffffffff,
                description: 'An integer between 0 and 4294967295',
            };
        case 'int64':
        case 'sint64':
        case 'sfixed64':
            return {
                type: ['integer', 'string', 'null'],
                minimum: Number(0n - 2n ** 63n), // This is not precise, due to JavaScript
                maximum: Number(2n ** 63n - 1n), // This is not precise, due to JavaScript
                pattern: '^[0-9]+$', // Can't enforce numeric limits on strings
                description: 'An integer between -9223372036854775808 and 9223372036854775807',
            };
        case 'uint64':
        case 'fixed64':
            return {
                type: ['integer', 'string', 'null'],
                minimum: 0,
                maximum: Number(2n ** 64n - 1n), // This is not precise, due to JavaScript
                pattern: '^[0-9]+$', // Can't enforce numeric limits on strings
                description: 'An integer between 0 and 18446744073709551615',
            };
        case 'float':
        case 'double':
            return {
                anyOf: [
                    { type: ['number', 'null'] },
                    { type: 'string', enum: ['NaN', 'Infinity', '-Infinity'] },
                ],
                description: 'A number, or one of the strings "NaN", "Infinity", or "-Infinity"',
            };
        default:
            console.warn(`Unrecognized Protobuf type: ${typeName}`);
            return {};
    }
}
// Convert a Protobuf runtime value into a canonical JSON value.
function normalizeProtobufValue(value) {
    if (protobufjs_1.default.util.Long.isLong(value)) {
        return value.toString(10);
    }
    if (value instanceof Uint8Array) {
        return js_base64_1.Base64.fromUint8Array(value);
    }
    if (Array.isArray(value)) {
        return js_base64_1.Base64.fromUint8Array(new Uint8Array(value));
    }
    return value;
}
