// @postman/runtime.models/dist/item-types/folder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Folder = void 0;
const z = __importStar(require("zod/v4-mini"));
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const events_js_1 = require("../extensions/events.js");
const item_type_js_1 = require("../item-type.js");
const graphql_request_js_1 = require("./graphql-request.js");
const grpc_request_js_1 = require("./grpc-request.js");
const http_request_js_1 = require("./http-request.js");
const llm_request_js_1 = require("./llm-request.js");
const mcp_request_js_1 = require("./mcp-request.js");
const mqtt_request_js_1 = require("./mqtt-request.js");
const ws_raw_request_js_1 = require("./ws-raw-request.js");
const ws_socketio_request_js_1 = require("./ws-socketio-request.js");
exports.Folder = (0, item_type_js_1.itemType)({
    type: 'folder',
    payload: z.strictObject({}),
    children: {
        allowed: [
            z.lazy(() => exports.Folder),
            ws_raw_request_js_1.WebSocketRequest,
            ws_socketio_request_js_1.SocketIORequest,
            grpc_request_js_1.GRPCRequest,
            graphql_request_js_1.GraphQLRequest,
            mqtt_request_js_1.MQTTRequest,
            llm_request_js_1.LLMRequest,
            http_request_js_1.HTTPRequest,
            mcp_request_js_1.MCPRequest,
        ],
    },
    extensions: {
        optional: {
            documentation: documentation_js_1.Documentation,
            events: events_js_1.Events.constraints({
                supportedListeners: [
                    'graphql:beforeQuery',
                    'graphql:afterResponse',
                    'grpc:beforeInvoke',
                    'grpc:onIncomingMessage',
                    'grpc:afterResponse',
                    'http:beforeRequest',
                    'http:afterResponse',
                ],
            }),
            auth: auth_js_1.Auth.constraints({
                supportedAuthTypes: [
                    'noauth',
                    'awsv4',
                    'basic',
                    'bearer',
                    'hawk',
                    'oauth1',
                    'oauth2',
                    'apikey',
                    'edgegrid',
                    'jwt',
                    'digest',
                    'ntlm',
                    'asap',
                    'inherit',
                ],
                isMultiAuthSupported: true,
            }),
        },
    },
});
