// @postman/runtime.models/dist/item-types/graphql-example.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLExample = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const schema_js_1 = require("../extensions/schema.js");
const item_type_js_1 = require("../item-type.js");
exports.GraphQLExample = (0, item_type_js_1.itemType)({
    type: 'graphql-example',
    payload: z.strictObject({
        url: z.string(),
        query: z.string(),
        variables: z.optional(z.string()),
        operationType: z.optional(z.enum(['query', 'mutation', 'subscription'])),
        operationName: z.optional(z.string()),
        headers: z.optional(key_value_js_1.DescriptiveKeyValues),
        response: z.optional(z.strictObject({
            http: z.optional(z.strictObject({
                status: z.string(),
                headers: key_value_js_1.KeyValues,
                body: z.optional(z.string()),
            })),
            events: z.array(z.union([
                z.strictObject({
                    id: z.string(),
                    type: z.enum(['received-response']),
                    content: z.string(),
                }),
            ])),
        })),
        settings: z.optional(z.strictObject({
            disabledSystemHeaders: z.optional(z.array(z.string())),
        })),
    }),
    children: {
        limit: 0,
    },
    extensions: {
        optional: {
            auth: auth_js_1.Auth.constraints({
                supportedAuthTypes: [
                    'noauth',
                    'awsv4',
                    'basic',
                    'bearer',
                    'hawk',
                    'oauth1',
                    'oauth2',
                    'apikey',
                    'edgegrid',
                    'inherit',
                ],
            }),
            documentation: documentation_js_1.Documentation,
            schema: schema_js_1.Schema,
        },
    },
});
