// @postman/runtime.models/dist/item-types/grpc-request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GRPCRequest = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const events_js_1 = require("../extensions/events.js");
const schema_js_1 = require("../extensions/schema.js");
const item_type_js_1 = require("../item-type.js");
const grpc_example_js_1 = require("./grpc-example.js");
exports.GRPCRequest = (0, item_type_js_1.itemType)({
    type: 'grpc-request',
    payload: z.strictObject({
        url: z.string(),
        methodPath: z.string(),
        methodDescriptor: z.optional(z.string()),
        message: z.strictObject({
            content: z.string(),
        }),
        metadata: key_value_js_1.DescriptiveKeyValues,
        settings: z.optional(z.strictObject({
            secureConnection: z.optional(z.boolean()),
            strictSSL: z.optional(z.boolean()),
            serverNameOverride: z.optional(z.string()),
            maxResponseMessageSize: z.optional(z.number()),
            includeDefaultFields: z.optional(z.boolean()),
            connectionTimeout: z.optional(z.number()),
        })),
    }),
    children: {
        allowed: [grpc_example_js_1.GRPCExample],
    },
    extensions: {
        optional: {
            auth: auth_js_1.Auth.constraints({
                supportedAuthTypes: [
                    'noauth',
                    'apikey',
                    'basic',
                    'bearer',
                    'oauth2',
                    'inherit',
                ],
            }),
            documentation: documentation_js_1.Documentation,
            events: events_js_1.Events.constraints({
                supportedListeners: [
                    'beforeInvoke',
                    'onIncomingMessage',
                    'afterResponse',
                ],
            }),
            schema: schema_js_1.Schema,
        },
    },
});
