// @postman/runtime.models/dist/item-types/http-example.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTTPExample = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
exports.HTTPExample = (0, item_type_js_1.itemType)({
    type: 'http-example',
    payload: z.strictObject({
        url: z.string(),
        method: z.string(),
        headers: z.optional(key_value_js_1.DescriptiveKeyValues),
        queryParams: z.optional(key_value_js_1.DescriptiveNullableKeyValues),
        pathVariables: z.optional(key_value_js_1.DescriptiveKeyValues),
        body: z.optional(z.discriminatedUnion('type', [
            z.strictObject({ type: z.literal('none') }),
            z.strictObject({
                type: z.literal('formdata'),
                content: z.array(z.discriminatedUnion('type', [
                    z.strictObject({
                        type: z.literal('text'),
                        key: z.string(),
                        value: z.string(),
                        contentType: z.optional(z.string()),
                        description: z.optional(z.string()),
                        disabled: z.optional(z.boolean()),
                    }),
                    z.strictObject({
                        type: z.literal('file'),
                        key: z.string(),
                        src: z.union([z.string(), z.array(z.string())]),
                        contentType: z.optional(z.string()),
                        description: z.optional(z.string()),
                        disabled: z.optional(z.boolean()),
                    }),
                ])),
            }),
            z.strictObject({
                type: z.literal('urlencoded'),
                content: key_value_js_1.DescriptiveKeyValues,
            }),
            z.strictObject({
                type: z.literal(['text', 'json', 'xml', 'html', 'javascript']),
                content: z.string(),
            }),
            z.strictObject({
                type: z.literal('file'),
                content: z.strictObject({
                    src: z.optional(z.string()),
                }),
            }),
        ])),
        response: z.optional(z.strictObject({
            statusCode: z.number(),
            statusText: z.string(),
            headers: z.optional(key_value_js_1.DescriptiveKeyValues),
            body: z.discriminatedUnion('type', [
                z.strictObject({
                    type: z.enum(['text', 'json', 'xml', 'html', 'javascript']),
                    content: z.string(),
                }),
            ]),
        })),
    }),
    children: {
        limit: 0,
    },
    extensions: {
        optional: {
            documentation: documentation_js_1.Documentation,
        },
    },
});
