// @postman/runtime.models/dist/item-types/llm-request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LLMRequest = void 0;
const z = __importStar(require("zod/v4-mini"));
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
const promptTypes = [
    'text',
    'image',
    'audio',
    'video',
    'postman_json',
    'json',
    'xml',
    'html',
    'binary',
    'hexdump',
];
exports.LLMRequest = (0, item_type_js_1.itemType)({
    type: 'llm-request',
    payload: z.strictObject({
        url: z.string(),
        config: z.strictObject({
            model: z.string(),
            provider: z.string(),
        }),
        settings: z.strictObject({
            temperature: z.optional(z.number()),
            presencePenalty: z.optional(z.number()),
            frequencyPenalty: z.optional(z.number()),
            maxToken: z.optional(z.int()),
            streamResponse: z.optional(z.boolean()),
            responseFormatJSON: z.optional(z.boolean()),
            topP: z.optional(z.number()),
            topK: z.optional(z.number()),
            repetitionPenalty: z.optional(z.number()),
            contextWindow: z.optional(z.number()),
            seed: z.optional(z.number()),
            maxSteps: z.optional(z.number()),
            streamTools: z.optional(z.boolean()),
        }),
        userPrompts: z.array(z.strictObject({
            id: z.optional(z.string()),
            value: z.optional(z.string()),
            timestamp: z.optional(z.number()),
            active: z.optional(z.boolean()),
            type: z.optional(z.enum(promptTypes)),
        })),
        systemPrompts: z.array(z.strictObject({
            id: z.optional(z.string()),
            value: z.optional(z.string()),
            timestamp: z.optional(z.number()),
            active: z.optional(z.boolean()),
            type: z.optional(z.enum(promptTypes)),
        })),
        mcpConfig: z.optional(z.string()),
        enabledTools: z.optional(z.array(z.string())),
    }),
    extensions: {
        optional: {
            auth: auth_js_1.Auth.constraints({
                supportedAuthTypes: [
                    'noauth',
                    'basic',
                    'bearer',
                    'apikey',
                    'inherit',
                    'awsv4',
                ],
            }),
            documentation: documentation_js_1.Documentation,
        },
    },
});
