// @postman/runtime.models/dist/item-types/mcp-request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPRequest = void 0;
const z = __importStar(require("zod/v4-mini"));
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
exports.MCPRequest = (0, item_type_js_1.itemType)({
    type: 'mcp-request',
    payload: z.discriminatedUnion('transport', [
        // SSE Transport
        z.strictObject({
            transport: z.literal('sse'),
            url: z.optional(z.string()),
            settings: z.optional(z.strictObject({
                strictSSL: z.optional(z.boolean()),
                requestTimeout: z.optional(z.number()),
                sessionTimeout: z.optional(z.number()),
            })),
            headers: z.optional(z.array(z.object({
                key: z.string(),
                value: z.string(),
                description: z.optional(z.string()),
                disabled: z.optional(z.boolean()),
            }))),
            message: z.optional(z.string()),
        }),
        // STDIO Transport
        z.strictObject({
            transport: z.literal('stdio'),
            command: z.optional(z.string()),
            env: z.optional(z.array(z.object({
                key: z.string(),
                value: z.string(),
                description: z.optional(z.string()),
                disabled: z.optional(z.boolean()),
            }))),
            settings: z.optional(z.strictObject({
                requestTimeout: z.optional(z.number()),
            })),
            message: z.optional(z.string()),
        }),
    ]),
    extensions: {
        optional: {
            auth: auth_js_1.Auth.constraints({
                supportedAuthTypes: ['noauth', 'basic', 'bearer', 'oauth2', 'inherit'],
            }),
            documentation: documentation_js_1.Documentation,
        },
    },
});
