// @postman/runtime.models/dist/item-types/mqtt-example.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MQTTExample = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
const MessageProperties = z.strictObject({
    payloadFormatIndicator: z.optional(z.boolean()),
    messageExpiryInterval: z.optional(z.int().check(z.minimum(0))),
    topicAlias: z.optional(z.int()),
    responseTopic: z.optional(z.string()),
    correlationData: z.optional(z.string()),
    subscriptionIdentifier: z.optional(z.int()),
    contentType: z.optional(z.string()),
    userProperties: z.optional(key_value_js_1.KeyValues),
});
exports.MQTTExample = (0, item_type_js_1.itemType)({
    type: 'mqtt-example',
    payload: z.strictObject({
        url: z.string(),
        clientId: z.optional(z.string()),
        version: z.optional(z.literal([4, 5])),
        topics: z.optional(z.array(z.strictObject({
            name: z.string(),
            qos: z.optional(z.literal([0, 1, 2])),
            subscribe: z.optional(z.boolean()),
            description: z.optional(z.string()),
            settings: z.optional(z.strictObject({
                noLocal: z.optional(z.boolean()),
                retainAsPublished: z.optional(z.boolean()),
                retainHandling: z.optional(z.literal([0, 1, 2])),
                subscriptionIdentifier: z.optional(z.int()),
            })),
            userProperties: z.optional(key_value_js_1.DescriptiveKeyValues),
        }))),
        lastWill: z.optional(z.strictObject({
            payload: z.optional(z.string()),
            topic: z.optional(z.string()),
            qos: z.optional(z.literal([0, 1, 2])),
            retain: z.optional(z.boolean()),
            type: z.optional(z.enum(['text', 'json', 'hex', 'base64'])),
            properties: z.optional(z.extend(MessageProperties, {
                willDelayInterval: z.optional(z.int().check(z.minimum(0))),
            })),
        })),
        properties: z.optional(z.strictObject({
            sessionExpiryInterval: z.optional(z.int().check(z.minimum(0))),
            receiveMaximum: z.optional(z.int().check(z.minimum(0))),
            maximumPacketSize: z.optional(z.int().check(z.minimum(0))),
            requestResponseInformation: z.optional(z.boolean()),
            userProperties: z.optional(key_value_js_1.DescriptiveKeyValues),
        })),
        events: z.optional(z.array(z.discriminatedUnion('type', [
            // Message events (incoming/outgoing)
            z.strictObject({
                id: z.string(),
                type: z.enum(['incoming', 'outgoing']),
                topic: z.optional(z.string()),
                content: z.optional(z.string()),
                qos: z.optional(z.literal([0, 1, 2])),
                retain: z.optional(z.boolean()),
                contentType: z.optional(z.enum(['text', 'json', 'binary'])),
                properties: z.optional(MessageProperties),
            }),
            // Error event
            z.strictObject({
                id: z.string(),
                type: z.literal('error'),
                code: z.optional(z.string()),
                content: z.optional(z.string()),
            }),
            // Info/Success events
            z.strictObject({
                id: z.string(),
                type: z.enum(['info', 'success']),
                content: z.optional(z.string()),
                generated: z.optional(z.boolean()),
            }),
        ]))),
    }),
    children: {
        limit: 0,
    },
    extensions: {
        optional: {
            auth: auth_js_1.Auth.constraints({
                supportedAuthTypes: ['noauth', 'basic', 'inherit'],
            }),
            documentation: documentation_js_1.Documentation,
        },
    },
});
