// @postman/runtime.models/dist/item-types/ws-raw-example.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketExample = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
exports.WebSocketExample = (0, item_type_js_1.itemType)({
    type: 'ws-raw-example',
    payload: z.strictObject({
        url: z.string(),
        queryParams: z.optional(key_value_js_1.DescriptiveNullableKeyValues),
        headers: z.optional(key_value_js_1.DescriptiveKeyValues),
        response: z.optional(z.strictObject({
            disconnectStatus: z.optional(z.string()),
            events: z.optional(z.array(z.discriminatedUnion('type', [
                // Message events (incoming/outgoing)
                z.strictObject({
                    type: z.enum(['incoming', 'outgoing']),
                    id: z.string(),
                    content: z.optional(z.string()),
                    contentType: z.optional(z.enum(['text', 'json', 'hex'])),
                }),
                // Error/Success events
                z.strictObject({
                    type: z.enum(['error', 'success']),
                    id: z.string(),
                    statusCode: z.optional(z.number()),
                    requestUrl: z.optional(z.string()),
                    generated: z.optional(z.boolean()),
                    responseHeaders: z.optional(key_value_js_1.DescriptiveKeyValues),
                }),
                // Info events
                z.strictObject({
                    type: z.literal('info'),
                    id: z.string(),
                    content: z.optional(z.string()),
                    generated: z.optional(z.boolean()),
                }),
            ]))),
        })),
        settings: z.optional(z.object()),
    }),
    children: {
        limit: 0,
    },
    extensions: {
        optional: {
            documentation: documentation_js_1.Documentation,
        },
    },
});
