// @postman/runtime.models/dist/item-types/ws-socketio-message.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketIOMessage = void 0;
const z = __importStar(require("zod/v4-mini"));
const item_type_js_1 = require("../item-type.js");
// Message argument schema
const Arg = z.discriminatedUnion('type', [
    // Text-based messages
    z.strictObject({
        content: z.string(),
        type: z.enum(['text', 'json']),
    }),
    // Binary messages
    z.strictObject({
        content: z.string(),
        type: z.literal('binary'),
        subtype: z.enum(['hex', 'base64']),
    }),
]);
exports.SocketIOMessage = (0, item_type_js_1.itemType)({
    type: 'ws-socketio-message',
    payload: z.strictObject({
        eventName: z.string(),
        acknowledgement: z.optional(z.boolean()),
        args: z.array(Arg).check(z.minLength(1)),
    }),
    children: {
        limit: 0,
    },
});
