// @postman/runtime.models/dist/item-types/ws-socketio-request.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketIORequest = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
const ws_socketio_example_js_1 = require("./ws-socketio-example.js");
const ws_socketio_message_js_1 = require("./ws-socketio-message.js");
exports.SocketIORequest = (0, item_type_js_1.itemType)({
    type: 'ws-socketio-request',
    payload: z.strictObject({
        url: z.optional(z.string()),
        headers: z.optional(key_value_js_1.DescriptiveKeyValues),
        queryParams: z.optional(key_value_js_1.DescriptiveNullableKeyValues),
        settings: z.optional(z.strictObject({
            version: z.optional(z.enum(['2', '3', '4'])),
            path: z.optional(z.string()),
            handshakeTimeout: z.optional(z.int().check(z.minimum(0))),
            retryCount: z.optional(z.int().check(z.minimum(0))),
            retryDelay: z.optional(z.int().check(z.minimum(0))),
            strictSSL: z.optional(z.boolean()),
        })),
        events: z.optional(z.array(z.strictObject({
            name: z.string(),
            description: z.optional(z.string()),
            subscribeOnConnect: z.optional(z.boolean()),
        }))),
    }),
    children: {
        allowed: [ws_socketio_message_js_1.SocketIOMessage, ws_socketio_example_js_1.SocketIOExample],
    },
    extensions: {
        optional: {
            documentation: documentation_js_1.Documentation,
        },
    },
});
