// @postman/runtime/dist/plugins/auth/asap/asap-auth-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.asapAuth = exports.ASAPAuthConfig = void 0;
const runtime_models_1 = require("@postman/runtime.models");
const jose_1 = require("jose");
const z = __importStar(require("zod/v4-mini"));
const arrays_1 = require("../../../lib/arrays");
const dates_1 = require("../../../lib/dates");
const key_values_1 = require("../../../lib/key-values");
const auth_1 = require("../auth");
exports.ASAPAuthConfig = z.object({
    alg: z.optional(z.string()),
    kid: z.optional(z.string()),
    iss: z.optional(z.string()),
    exp: z.optional(z.string()),
    aud: z.optional(z.string()),
    sub: z.optional(z.string()),
    privateKey: z.optional(z.string()),
    claims: z.optional(z.union([
        z.string(),
        z.object({
            kid: z.optional(z.string()),
            iss: z.optional(z.string()),
            exp: z.optional(z.number()),
            aud: z.optional(z.string()),
            sub: z.optional(z.string()),
            jti: z.optional(z.string()),
            iat: z.optional(z.number()),
        }),
    ])),
});
const DEFAULT_EXPIRY = 3600;
const SUPPORTED_ALGORITHMS = [
    'RS256',
    'RS384',
    'RS512',
    'PS256',
    'PS384',
    'PS512',
    'ES256',
    'ES384',
    'ES512',
];
exports.asapAuth = {
    name: 'asap-auth',
    resolved: {
        async 'http:send-request'(event, run) {
            if (!runtime_models_1.HTTPRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'asap');
            if (!auth)
                return;
            const result = exports.ASAPAuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.asap ?? []));
            if (!result.success)
                return;
            const config = result.data;
            let { alg = 'RS256', privateKey, claims = {} } = config;
            if (typeof claims === 'string') {
                try {
                    claims = JSON.parse(claims.trim() || '{}');
                }
                catch (cause) {
                    throw new Error('Invalid ASAP configuration, failed to parse claims', {
                    // TODO cause
                    });
                }
            }
            // Give priority to claims, if present
            const issuer = claims.iss ?? config.iss;
            const kid = claims.kid ?? config.kid;
            // SHOULD: fall back to issuer, if not present
            const subject = claims.sub ?? config.sub ?? issuer;
            const audience = claims.aud ?? config.aud;
            // MUST: token id is UUID
            const jwtTokenId = claims.jti ?? crypto.randomUUID();
            const currentTimestamp = (0, dates_1.getTimestampInSeconds)(new Date());
            const issuedAt = claims.iat ?? currentTimestamp;
            const expiry = config.exp && !isNaN(parseInt(config.exp, 10)) ?
                parseInt(config.exp, 10)
                : DEFAULT_EXPIRY;
            const expiryTimestamp = claims.exp ?? currentTimestamp + expiry;
            if (!kid || !issuer || !audience || !jwtTokenId || !privateKey) {
                throw new Error('Invalid ASAP configuration, missing one or more required claims');
            }
            if (!SUPPORTED_ALGORITHMS.includes(alg)) {
                throw new Error(`Invalid ASAP configuration, unsupported algorithm`);
            }
            if (typeof privateKey !== 'string') {
                throw new Error('Invalid ASAP configuration, private key must be a string');
            }
            const signKey = await (0, jose_1.importPKCS8)(parsePrivateKey(privateKey, kid), alg);
            const token = await new jose_1.SignJWT(claims)
                .setProtectedHeader({ alg, kid })
                .setIssuedAt(issuedAt)
                .setIssuer(issuer)
                .setSubject(subject)
                .setJti(jwtTokenId)
                .setAudience(audience)
                .setExpirationTime(expiryTimestamp)
                .sign(signKey);
            run.patch(event, (payload) => {
                (0, arrays_1.filter)(payload.headers, (header) => header.key.toLowerCase() !== 'authorization');
                payload.headers.push({
                    key: 'authorization',
                    value: `Bearer ${token}`,
                    system: true,
                });
            });
        },
    },
};
const DATA_URI_PATTERN = /^data:application\/pkcs8;kid=([\w.\-\+/]+);base64,([a-zA-Z0-9+/=]+)$/;
function parsePrivateKey(value, keyId) {
    // Remove newlines and double-quotes in key
    const prepared = value.replace(/\n/g, '').replace(/^"(.*)"$/, '$1');
    const uriDecodedPrivateKey = decodeURIComponent(prepared);
    if (!uriDecodedPrivateKey.startsWith('data:')) {
        return uriDecodedPrivateKey;
    }
    const match = DATA_URI_PATTERN.exec(uriDecodedPrivateKey);
    if (!match) {
        throw new Error('Unable to parse private key, malformed data URI');
    }
    const [_, matchedKeyId, matchedData] = match;
    if (keyId !== matchedKeyId) {
        throw new Error('Unable to parse private key, supplied key id does not match the one included in data uri.');
    }
    // TODO base64 -> PEM
    return matchedData;
}
