// @postman/runtime.models/dist/common/key-value.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescriptiveNullableKeyValues = exports.NullableKeyValues = exports.DescriptiveUnknownKeyValues = exports.UnknownKeyValues = exports.DescriptiveKeyValues = exports.KeyValues = void 0;
const z = __importStar(require("zod/v4-mini"));
const metadata_1 = require("../metadata");
const model_1 = require("../model");
exports.KeyValues = (0, model_1.model)(z
    .array(z.strictObject({
    key: z.string(),
    value: z.string(),
}))
    .register(metadata_1.metadata, { id: 'key-values' }));
exports.DescriptiveKeyValues = (0, model_1.model)(z
    .array(z.strictObject({
    key: z.string(),
    value: z.string(),
    description: z.optional(z.string()),
    disabled: z.optional(z.boolean()),
}))
    .register(metadata_1.metadata, { id: 'descriptive-key-values' }));
exports.UnknownKeyValues = (0, model_1.model)(z
    .array(z.strictObject({
    key: z.string(),
    value: z.unknown(),
}))
    .register(metadata_1.metadata, { id: 'unknown-key-values' }));
exports.DescriptiveUnknownKeyValues = (0, model_1.model)(z
    .array(z.strictObject({
    key: z.string(),
    value: z.unknown(),
    description: z.optional(z.string()),
    disabled: z.optional(z.boolean()),
}))
    .register(metadata_1.metadata, { id: 'descriptive-unknown-key-values' }));
exports.NullableKeyValues = (0, model_1.model)(z
    .array(z.strictObject({
    key: z.nullable(z.string()),
    value: z.nullable(z.string()),
}))
    .register(metadata_1.metadata, { id: 'nullable-key-values' }));
exports.DescriptiveNullableKeyValues = (0, model_1.model)(z
    .array(z.strictObject({
    key: z.nullable(z.string()),
    value: z.nullable(z.string()),
    description: z.optional(z.string()),
    disabled: z.optional(z.boolean()),
}))
    .register(metadata_1.metadata, { id: 'descriptive-nullable-key-values' }));
