// @postman/runtime/dist/lib/rpc/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvokeOptions = exports.Channel = void 0;
exports.invoke = invoke;
const z = __importStar(require("zod/v4-mini"));
/**
 * Channel is a bi-directional stream of incoming and outgoing values.
 * It can represent values over a WebSocket, IPC, or other connections.
 * It's based on the design of [WebTransport][1] and other bidirection / duplex streams
 *
 * [1]: https://developer.mozilla.org/en-US/docs/Web/API/WebTransportBidirectionalStream
 *
 * @example
 * ```ts
 * function connect() {
 *   return new Channel(async ({ incoming, outgoing }) => {
 *     const socket = new WebSocket('...');
 *     socket.addEventListener('message', (data) => {
 *       incoming.enqueue(data);
 *     });
 *     socket.addEventListener('error', (error) => {
 *       incoming.error(error);
 *     });
 *     socket.addEventListener('close', () => {
 *       incoming.close();
 *     })
 *
 *     // Send all writes to socket
 *     for await (const message of outgoing) {
 *       socket.send(message);
 *     }
 *
 *     // Close socket when writer closes
 * 		 socket.close()
 *   });
 * }
 *
 * // Echo messages back to WebSocket
 * const channel = connect();
 * const writer = channel.writable.getWriter();
 * for await (const message of channel.readable) {
 *   if (message === 'quit') {
 *     writer.close();
 *     break;
 *   }
 *
 *   writer.write(message);
 * }
 * ```
 */
class Channel {
    readable;
    writable;
    constructor(
    /**
     * Handler receives the "other end" of the channel,
     * with `incoming` writing to the channel's readable stream
     * and `outgoing` reading from the channel's writable stream
     */
    handler) {
        // Outgoing
        // (from channel.writable -> handler)
        let outgoingController;
        let writableController;
        const outgoing = new ReadableStream({
            start(controller) {
                outgoingController = controller;
            },
            cancel(reason) {
                // When `outgoing` is cancelled in handler,
                // `writable` should not accept any more writes
                //
                // From [WebTransport][1] spec, should error stream
                // [1]: https://www.w3.org/TR/webtransport/#send-stream-STOP_SENDING
                writableController?.error(reason);
            },
        });
        this.writable = new WritableStream({
            start(controller) {
                writableController = controller;
            },
            write(data) {
                outgoingController.enqueue(data);
            },
            abort() {
                // When aborting writable, may make more sense to cancel outgoing,
                // but likely already written so close should be sufficient
                //
                // `cancel` would likely cause locked errors on outgoing directly
                try {
                    outgoingController.close();
                }
                catch (_error) {
                    // (ignore already closed errors)
                }
            },
            close() {
                try {
                    outgoingController.close();
                }
                catch (_error) {
                    // (ignore already closed errors)
                }
            },
        });
        // Incoming
        // (from handler -> channel.readable)
        let incomingController;
        const readableCancellationController = new AbortController();
        this.readable = new ReadableStream({
            start(controller) {
                incomingController = controller;
            },
            cancel() {
                readableCancellationController.abort();
            },
        });
        const incoming = Object.assign(incomingController, {
            signal: readableCancellationController.signal,
        });
        (async () => {
            await handler({ incoming, outgoing });
        })().catch((error) => {
            try {
                writableController.error(error);
                outgoingController.error(error);
                incomingController.error(error);
            }
            catch (_error) {
                // (ignore invalid state errors)
            }
        });
    }
}
exports.Channel = Channel;
exports.InvokeOptions = z.object({
    path: z.array(z.string()),
    args: z.array(z.any()),
});
const ChannelSchema = z.object({
    readable: z.unknown(),
    writable: z.unknown(),
});
async function invoke(platform, options) {
    const optionsResult = exports.InvokeOptions.safeParse(options);
    if (!optionsResult.success) {
        throw new Error('Invalid invocation: invalid options');
    }
    const { path, args } = optionsResult.data;
    const [namespace, name] = path;
    const method = platform[namespace]?.[name];
    if (!method) {
        throw new Error(`Invalid invocation: method not found for ${JSON.stringify(path)}`);
    }
    const channel = await method(...args);
    const channelResult = ChannelSchema.safeParse(channel);
    if (!channelResult.success) {
        throw new Error('Invalid invocation: unexpected result');
    }
    return channel;
}
