// @postman/runtime/dist/plugins/scripting/utils/event-emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventEmitter = void 0;
class EventEmitter {
    handlers = {};
    addEventListener(event, handler, opts) {
        if (opts?.signal && opts.signal.aborted) {
            return;
        }
        let eventHandlers = this.handlers[event];
        if (!eventHandlers) {
            eventHandlers = [];
            this.handlers[event] = eventHandlers;
        }
        let h = opts?.once ? wrapOnce(this, event, handler) : handler;
        if (opts?.signal) {
            const signalFn = () => {
                this.removeEventListener(event, h);
            };
            opts?.signal?.addEventListener('abort', signalFn, { once: true });
            h = Object.assign(h, { signal: opts.signal, signalFn });
        }
        this.handlers[event].push(h);
    }
    removeEventListener(event, handler) {
        const eventHandlers = this.handlers[event];
        if (!handler) {
            return;
        }
        const index = eventHandlers.findIndex((h) => h === handler || h.listener === handler);
        if (index === -1) {
            return;
        }
        let h = eventHandlers[index];
        if (h.signal && h.signalFn) {
            h.signal.removeEventListener('abort', h.signalFn);
        }
        this.handlers[event].splice(index, 1);
    }
    offAll() {
        this.handlers = {};
    }
    emit(event, arg) {
        const eventHandlers = this.handlers[event];
        if (!eventHandlers) {
            return;
        }
        eventHandlers.forEach((handler) => {
            handler(arg);
        });
    }
}
exports.EventEmitter = EventEmitter;
function onceWrapper(arg) {
    if (!this.fired) {
        this.target.removeEventListener(this.type, this.wrapFn);
        this.fired = true;
        this.listener(arg);
    }
}
function wrapOnce(target, type, listener) {
    const state = {
        fired: false,
        target,
        type,
        listener,
        wrapFn: () => { },
    };
    const wrapped = Object.assign(onceWrapper.bind(state), { listener });
    state.wrapFn = wrapped;
    return wrapped;
}
