// @postman/grpc-reflection-js/build/src/client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const grpc_js_1 = require("@postman/grpc-js");
const descriptor_1 = require("./descriptor");
const descriptor_2 = require("@postman/protobufjs/ext/descriptor");
const lodash_set_1 = __importDefault(require("lodash.set"));
const supportedReflectionAPIVersions = {
    v1alpha: {
        priority: 0,
        serviceName: 'grpc.reflection.v1alpha.ServerReflection',
        client: Promise.resolve().then(() => __importStar(require('./reflection_providers/v1alpha/reflection_pb'))),
        service: Promise.resolve().then(() => __importStar(require('./reflection_providers/v1alpha/reflection_grpc_pb'))),
    },
    v1: {
        priority: 1,
        serviceName: 'grpc.reflection.v1.ServerReflection',
        client: Promise.resolve().then(() => __importStar(require('./reflection_providers/v1/reflection_pb'))),
        service: Promise.resolve().then(() => __importStar(require('./reflection_providers/v1/reflection_grpc_pb'))),
    },
};
class Client {
    constructor(url, credentials, options, metadata) {
        this.fileDescriptorCache = new Map();
        this.url = url;
        this.credentials = credentials;
        this.clientOptions = options;
        this.fileDescriptorCache = new Map();
        this.metadata = metadata || new grpc_js_1.Metadata();
    }
    async sendReflectionRequest(request, client) {
        return new Promise((resolve, reject) => {
            const result = [];
            const grpcCall = (client || this.grpcClient).serverReflectionInfo(this.metadata);
            grpcCall.on('data', (response) => {
                result.push(response);
            });
            grpcCall.on('error', (error) => {
                reject(error);
            });
            grpcCall.on('end', () => resolve(result));
            if (Array.isArray(request)) {
                request.forEach(req => grpcCall.write(req));
            }
            else {
                grpcCall.write(request);
            }
            grpcCall.end();
        });
    }
    async evaluateSupportedServerReflectionProtocol() {
        const evaluationPromises = [];
        // Check version compatibility and initialize gRPC client based on that
        for (const version of Object.keys(supportedReflectionAPIVersions)) {
            evaluationPromises.push(
            // eslint-disable-next-line no-async-promise-executor
            new Promise(async (resolve) => {
                const protocolConfig = supportedReflectionAPIVersions[version];
                const { service: servicePromise, client: clientPromise, } = protocolConfig;
                const [protocolService, protocolClient] = await Promise.all([
                    servicePromise,
                    clientPromise,
                ]);
                const grpcClientForProtocol = new protocolService.ServerReflectionClient(this.url, this.credentials, this.clientOptions);
                const request = new protocolClient.ServerReflectionRequest();
                request.setListServices('*');
                try {
                    const [reflectionResponse] = await this.sendReflectionRequest(request, grpcClientForProtocol);
                    return resolve({
                        successful: true,
                        priority: protocolConfig.priority,
                        effect: () => {
                            this.grpcClient = grpcClientForProtocol;
                            this.compatibleReflectionVersion = version;
                            this.CompatibleServerReflectionRequest =
                                protocolClient.ServerReflectionRequest;
                            this.reflectionResponseCache = reflectionResponse;
                        },
                    });
                }
                catch (error) {
                    return resolve({
                        successful: false,
                        priority: protocolConfig.priority,
                        error: error,
                    });
                }
            }));
        }
        const evaluationResults = await Promise.all(evaluationPromises);
        const [successfulReflectionByPriority] = evaluationResults
            .filter(res => res.successful)
            .sort((res1, res2) => res2.priority - res1.priority);
        if (!successfulReflectionByPriority) {
            const reflectionNotImplementedError = evaluationResults.find(res => {
                return res.error && res.error.code === grpc_js_1.status.UNIMPLEMENTED;
            });
            const resultWithServiceError = evaluationResults.find(res => {
                // Something is actually wrong with the gRPC service
                return res.error && res.error.code !== grpc_js_1.status.UNIMPLEMENTED;
            });
            throw ((resultWithServiceError === null || resultWithServiceError === void 0 ? void 0 : resultWithServiceError.error) || (reflectionNotImplementedError === null || reflectionNotImplementedError === void 0 ? void 0 : reflectionNotImplementedError.error) ||
                new Error('No compatible reflection API found.'));
        }
        // Set grpc client and other properties based on highest priority successful version
        successfulReflectionByPriority.effect();
    }
    async initialize() {
        if (this.grpcClient || this.compatibleReflectionVersion)
            return;
        await this.evaluateSupportedServerReflectionProtocol();
    }
    async listServices() {
        await this.initialize();
        return new Promise((resolve, reject) => {
            function dataCallback(response) {
                var _a;
                if (response.hasListServicesResponse()) {
                    const services = (_a = response
                        .getListServicesResponse()) === null || _a === void 0 ? void 0 : _a.getServiceList().map(svc => {
                        return svc.getName();
                    });
                    resolve(services || []);
                }
                else {
                    reject(Error());
                }
            }
            function errorCallback(e) {
                reject(e);
            }
            if (this.reflectionResponseCache) {
                return dataCallback(this.reflectionResponseCache);
            }
            const request = new this.CompatibleServerReflectionRequest();
            request.setListServices('*');
            this.sendReflectionRequest(request)
                .then(([response]) => dataCallback(response))
                .catch(errorCallback);
        });
    }
    fileContainingSymbol(symbol) {
        return new Promise((resolve, reject) => {
            this.getFileContainingSymbol(symbol)
                .then(val => resolve(this.resolveFileDescriptorSet(val)))
                .catch(err => reject(err));
        });
    }
    fileByFilename(filename) {
        return new Promise((resolve, reject) => {
            this.getFilesByFilenames([filename])
                .then(val => resolve(this.resolveFileDescriptorSet(val)))
                .catch(err => reject(err));
        });
    }
    async resolveFileDescriptorSet(fileDescriptorProtos) {
        const fileDescriptorMap = await this.resolveDescriptorRecursive(fileDescriptorProtos);
        const fileDescriptorSet = descriptor_2.FileDescriptorSet.create();
        lodash_set_1.default(fileDescriptorSet, 'file', Array.from(fileDescriptorMap.values()));
        return descriptor_1.getDescriptorRootFromDescriptorSet(fileDescriptorSet);
    }
    async resolveDescriptorRecursive(fileDescriptorProtos = [], fileDescriptorMap = new Map()) {
        await Promise.all(fileDescriptorProtos.map(async (fileDescriptorProto) => {
            if (fileDescriptorMap.has(fileDescriptorProto.name)) {
                return;
            }
            else {
                fileDescriptorMap.set(fileDescriptorProto.name, fileDescriptorProto);
            }
            const dependencies = (fileDescriptorProto.dependency || []).filter((dependency) => !fileDescriptorMap.has(dependency));
            if (dependencies.length) {
                await this.resolveDescriptorRecursive(await this.getFilesByFilenames(dependencies), fileDescriptorMap);
            }
        }));
        return fileDescriptorMap;
    }
    async getFileContainingSymbol(symbol) {
        await this.initialize();
        const fileDescriptorCache = this.fileDescriptorCache;
        return new Promise((resolve, reject) => {
            function dataCallback(response) {
                var _a;
                if (response.hasFileDescriptorResponse()) {
                    const fileDescriptorProtoBytes = (((_a = response
                        .getFileDescriptorResponse()) === null || _a === void 0 ? void 0 : _a.getFileDescriptorProtoList()) || []);
                    resolve(fileDescriptorProtoBytes.map(descriptorByte => {
                        const fileDescriptorProto = descriptor_2.FileDescriptorProto.decode(descriptorByte);
                        fileDescriptorCache.set(fileDescriptorProto.name, fileDescriptorProto);
                        return fileDescriptorProto;
                    }));
                }
                else {
                    reject(Error());
                }
            }
            function errorCallback(e) {
                reject(e);
            }
            const request = new this.CompatibleServerReflectionRequest();
            request.setFileContainingSymbol(symbol);
            this.sendReflectionRequest(request)
                .then(([response]) => dataCallback(response))
                .catch(errorCallback);
        });
    }
    async getFilesByFilenames(symbols) {
        await this.initialize();
        const result = [];
        const fileDescriptorCache = this.fileDescriptorCache;
        const symbolsToFetch = symbols.filter(symbol => {
            const cached = fileDescriptorCache.get(symbol);
            if (cached) {
                result.push(cached);
                return false;
            }
            return true;
        });
        if (symbolsToFetch.length === 0) {
            return Promise.resolve(result);
        }
        return new Promise((resolve, reject) => {
            function dataCallback(response) {
                var _a, _b;
                if (response.hasFileDescriptorResponse()) {
                    (_b = (_a = response
                        .getFileDescriptorResponse()) === null || _a === void 0 ? void 0 : _a.getFileDescriptorProtoList()) === null || _b === void 0 ? void 0 : _b.forEach(descriptorByte => {
                        if (descriptorByte instanceof Uint8Array) {
                            const fileDescriptorProto = descriptor_2.FileDescriptorProto.decode(descriptorByte);
                            fileDescriptorCache.set(fileDescriptorProto.name, fileDescriptorProto);
                            result.push(fileDescriptorProto);
                        }
                    });
                }
                else if (response.hasErrorResponse()) {
                    const err = response.getErrorResponse();
                    reject(new Error(`Error: ${err === null || err === void 0 ? void 0 : err.getErrorCode()}: ${err === null || err === void 0 ? void 0 : err.getErrorMessage()}`));
                }
                else {
                    reject(Error());
                }
            }
            function errorCallback(e) {
                reject(e);
            }
            const requests = symbolsToFetch.map(symbol => {
                const request = new this.CompatibleServerReflectionRequest();
                return request.setFileByFilename(symbol);
            });
            this.sendReflectionRequest(requests)
                .then(responses => {
                for (const dataBit of responses)
                    dataCallback(dataBit);
                resolve(result);
            })
                .catch(errorCallback);
        });
    }
}
exports.Client = Client;
