// @postman/runtime/dist/plugins/auth/hawk/hawk.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateHawkAuthorization = generateHawkAuthorization;
const Hawk = __importStar(require("postman-request/lib/hawk"));
const hashing_1 = require("../../../lib/crypto/hashing");
const urls_1 = require("../../../lib/urls");
function generateHawkAuthorization(parameters) {
    const { url, method, hash, credentials, nonce, timestamp, extraData, app, delegation, user, } = parameters;
    const parts = (0, urls_1.parseURL)(url);
    const authorization = Hawk.header({
        ...parts,
        port: parts.port ? parseInt(parts.port) : undefined,
    }, method, {
        credentials: {
            ...credentials,
            algorithm: (0, hashing_1.toNodeHashAlgorithm)(credentials.algorithm),
        },
        nonce,
        timestamp,
        ext: extraData,
        app,
        dlg: delegation,
        user,
        url,
        method,
        hash,
    });
    return authorization;
}
