// @postman/runtime/dist/plugins/auth/hawk/hawk-auth-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.hawkAuth = exports.HawkAuthConfig = void 0;
const runtime_models_1 = require("@postman/runtime.models");
const z = __importStar(require("zod/v4-mini"));
const http_body_1 = require("../../../items/http/http-body");
const arrays_1 = require("../../../lib/arrays");
const hashing_1 = require("../../../lib/crypto/hashing");
const randomness_1 = require("../../../lib/crypto/randomness");
const dates_1 = require("../../../lib/dates");
const key_values_1 = require("../../../lib/key-values");
const auth_1 = require("../auth");
const hawk_1 = require("./hawk");
exports.HawkAuthConfig = z.object({
    authId: z.optional(z.string()),
    authKey: z.optional(z.string()),
    algorithm: z.optional(hashing_1.HashAlgorithm),
    nonce: z.optional(z.string()),
    timestamp: z.optional(z.string()),
    extraData: z.optional(z.string()),
    app: z.optional(z.string()),
    delegation: z.optional(z.string()),
    user: z.optional(z.string()),
    includePayloadHash: z.optional(z.boolean()),
});
exports.hawkAuth = {
    name: 'hawk-auth',
    resolved: {
        async 'http:send-request'(event, run) {
            if (!runtime_models_1.HTTPRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'hawk');
            if (!auth)
                return;
            const result = exports.HawkAuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.hawk ?? []));
            if (!result.success)
                return;
            const config = result.data;
            const { authId, authKey, algorithm = 'SHA-256', nonce = (0, randomness_1.randomASCII)(6), timestamp = `${(0, dates_1.getTimestampInSeconds)(new Date())}`, extraData = '', app, delegation, user, includePayloadHash, } = config;
            await run.patch(event, async (payload) => {
                (0, arrays_1.filter)(payload.headers, (header) => header.key.toLowerCase() !== 'authorization');
                if (!authId || !authKey)
                    return;
                const hash = includePayloadHash ?
                    await (0, http_body_1.computeBodyHash)(payload, 'SHA-256', 'base64', run.runtime.platform)
                    : undefined;
                const authorization = (0, hawk_1.generateHawkAuthorization)({
                    credentials: {
                        id: authId,
                        key: authKey,
                        algorithm,
                    },
                    nonce,
                    timestamp,
                    extraData,
                    app,
                    delegation,
                    user,
                    url: payload.url,
                    method: payload.method,
                    hash,
                });
                payload.headers.push({
                    key: 'authorization',
                    value: authorization,
                    system: true,
                });
            });
        },
    },
};
