// @postman/runtime/dist/plugins/auth/bearer/bearer-auth-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.bearerAuth = exports.BearerAuthConfig = void 0;
const runtime_models_1 = require("@postman/runtime.models");
const z = __importStar(require("zod/v4-mini"));
const key_values_1 = require("../../../lib/key-values");
const auth_1 = require("../auth");
exports.BearerAuthConfig = z.object({
    token: z.optional(z.string()),
});
exports.bearerAuth = {
    name: 'bearer-auth',
    resolved: {
        'http:send-request'(event, run) {
            if (!runtime_models_1.HTTPRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'bearer');
            if (!auth)
                return;
            const result = exports.BearerAuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.bearer));
            if (!result.success)
                return;
            const authorization = generateBearerAuthorization(result.data);
            run.patch(event, (payload) => {
                payload.headers.push({
                    key: 'Authorization',
                    value: authorization,
                    system: true,
                });
            });
        },
        'mcp:request-connect'(event, run) {
            if (!runtime_models_1.MCPRequest.is(run.item) || event.payload.transport !== 'sse')
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'bearer');
            if (!auth)
                return;
            const result = exports.BearerAuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.bearer));
            if (!result.success)
                return;
            const authorization = generateBearerAuthorization(result.data);
            run.patch(event, (payload) => {
                if (payload.transport === 'sse') {
                    payload.headers ??= [];
                    (0, key_values_1.upsertKeyValue)(payload.headers, {
                        key: 'Authorization',
                        value: authorization,
                    });
                }
            });
        },
        'grpc:invoke'(event, run) {
            if (!runtime_models_1.GRPCRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'bearer');
            if (!auth)
                return;
            const result = exports.BearerAuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.bearer));
            if (!result.success)
                return;
            const authorization = generateBearerAuthorization(result.data);
            run.patch(event, (payload) => {
                payload.metadata ??= [];
                (0, key_values_1.upsertKeyValue)(payload.metadata, { key: 'Authorization', value: authorization }, { caseInsensitive: true });
            });
        },
    },
};
function generateBearerAuthorization(parameters) {
    const { token = '' } = parameters;
    const authorization = `Bearer ${token}`;
    return authorization;
}
