// @postman/runtime/dist/plugins/auth/awsv4/awsv4-auth-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.awsv4Auth = exports.AWSV4AuthConfig = void 0;
const runtime_models_1 = require("@postman/runtime.models");
const aws4fetch_1 = require("aws4fetch");
const sdk = __importStar(require("postman-collection"));
const z = __importStar(require("zod/v4-mini"));
const http_body_1 = require("../../../items/http/http-body");
const arrays_1 = require("../../../lib/arrays");
const key_values_1 = require("../../../lib/key-values");
const auth_1 = require("../auth");
exports.AWSV4AuthConfig = z.object({
    accessKey: z.optional(z.string()),
    secretKey: z.optional(z.string()),
    sessionToken: z.optional(z.string()),
    service: z.optional(z.string()),
    region: z.optional(z.string()),
    addAuthDataToQuery: z.optional(z.boolean()),
});
exports.awsv4Auth = {
    name: 'awsv4-auth',
    resolved: {
        async 'http:send-request'(event, run) {
            if (!runtime_models_1.HTTPRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'awsv4');
            if (!auth)
                return;
            const config = exports.AWSV4AuthConfig.parse((0, key_values_1.keyValuesToObject)(auth.awsv4 ?? []));
            const { accessKey: accessKeyId, secretKey: secretAccessKey, sessionToken, region = 'us-east-1', service = 'execute-api', addAuthDataToQuery: signQuery = false, } = config;
            await run.patch(event, async (payload) => {
                // Assumption: This plugin takes control of the authorization flow,
                // remove user-defined Authorization and X-Amz headers / query params
                (0, arrays_1.filter)(payload.headers, (header) => ![
                    'authorization',
                    'x-amz-data',
                    'x-amz-security-token',
                    'x-amz-content-sha256',
                ].includes(header.key.toLowerCase()));
                (0, arrays_1.filter)(payload.queryParams, (param) => !param.key ||
                    ![
                        'x-amz-security-token',
                        'x-amz-date',
                        'x-amz-algorithm',
                        'x-amz-credential',
                        'x-amz-signedheaders',
                        'x-amz-signature',
                    ].includes(param.key.toLowerCase()));
                // Body hash is not required when adding auth data to query params
                // @see: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
                //
                // aws4fetch module can't calculate body hash for body with ReadableStream.
                // Calculate it and set 'x-amz-content-sha256' header
                // which will be used by aws4fetch module to calculate the signature.
                const bodyHash = !signQuery ?
                    await (0, http_body_1.computeBodyHash)(payload, 'SHA-256', 'hex', run.runtime.platform)
                    : undefined;
                if (bodyHash) {
                    payload.headers.push({
                        key: 'x-amz-content-sha256',
                        value: bodyHash,
                        system: true,
                    });
                }
                const headers = (0, key_values_1.keyValuesToObject)(payload.headers, {
                    approach: 'combine',
                });
                // Convert headers to HeadersInit format (array or object with string values)
                const headersInit = {};
                for (const [key, value] of Object.entries(headers)) {
                    headersInit[key] = Array.isArray(value) ? value.join(', ') : value;
                }
                const signer = new aws4fetch_1.AwsV4Signer({
                    url: payload.url,
                    accessKeyId: accessKeyId ?? '',
                    secretAccessKey: secretAccessKey ?? '',
                    sessionToken,
                    method: payload.method,
                    headers: headersInit,
                    // no need to give body since we are setting 'x-amz-content-sha256' header
                    body: undefined,
                    signQuery,
                    service,
                    region,
                    // ensure path encoding matches expectations in tests
                    singleEncode: true,
                });
                const signedRequest = await signer.sign();
                if (signQuery) {
                    const signedUrl = signedRequest.url;
                    const parsed = sdk.Url.parse(String(signedUrl));
                    const signedQueryParams = (parsed.query ?? []);
                    for (const param of signedQueryParams) {
                        if (param.key.toLowerCase().startsWith('x-amz-')) {
                            payload.queryParams.push({ ...param });
                        }
                    }
                }
                signedRequest.headers.forEach((value, key) => {
                    (0, key_values_1.upsertKeyValue)(payload.headers, {
                        key,
                        value: String(value ?? ''),
                        system: true,
                    });
                });
            });
        },
    },
};
