// @postman/runtime/dist/events.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeEventContext = exports.ISO8601 = void 0;
exports.isAnyEvent = isAnyEvent;
exports.isRuntimeEventContext = isRuntimeEventContext;
exports.isRuntimeEvent = isRuntimeEvent;
exports.isEventOfType = isEventOfType;
exports.isRuntimeEventOfType = isRuntimeEventOfType;
exports.isErrorEvent = isErrorEvent;
const z = __importStar(require("zod/v4-mini"));
const resources_1 = require("./lib/resources");
const schemas_1 = require("./lib/schemas");
function isAnyEvent(input) {
    return (input != null &&
        typeof input === 'object' &&
        'type' in input &&
        typeof input.type === 'string');
}
exports.ISO8601 = z.iso.datetime();
exports.RuntimeEventContext = z.object({
    timestamp: exports.ISO8601,
    source: z.string(),
    item: z.nullable(resources_1.Link),
    collection: z.nullable(resources_1.Link),
    folders: z.array(resources_1.Link),
    workspace: z.nullable(resources_1.Link),
    environment: z.nullable(resources_1.Link),
    run: resources_1.Link,
    process: z.array(resources_1.Link),
});
function isRuntimeEventContext(input) {
    return (0, schemas_1.is)(exports.RuntimeEventContext, input);
}
const RuntimeEvent = z.object({
    type: z.string(),
    id: z.string(),
    context: exports.RuntimeEventContext,
});
function isRuntimeEvent(input) {
    return (0, schemas_1.is)(RuntimeEvent, input);
}
function isEventOfType(type, event) {
    if (event === undefined) {
        return (event) => event.type === type;
    }
    return event.type === type;
}
function isRuntimeEventOfType(type, event) {
    if (event === undefined) {
        return (event) => event.type === type;
    }
    return event.type === type;
}
function isErrorEvent(input) {
    return (isAnyEvent(input) &&
        input.payload != null &&
        typeof input.payload === 'object' &&
        'error' in input.payload);
}
