// @postman/runtime/dist/next.node.js
"use strict";
/// <reference types="node" />
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runtime = exports.phases = exports.Run = exports.isRuntimeEventOfType = exports.isRuntimeEventContext = exports.isRuntimeEvent = exports.isEventOfType = exports.isErrorEvent = exports.isAnyEvent = void 0;
const platform_node_js_1 = require("./platform.node.js");
const runtime_js_1 = require("./runtime.js");
var events_js_1 = require("./events.js");
Object.defineProperty(exports, "isAnyEvent", { enumerable: true, get: function () { return events_js_1.isAnyEvent; } });
Object.defineProperty(exports, "isErrorEvent", { enumerable: true, get: function () { return events_js_1.isErrorEvent; } });
Object.defineProperty(exports, "isEventOfType", { enumerable: true, get: function () { return events_js_1.isEventOfType; } });
Object.defineProperty(exports, "isRuntimeEvent", { enumerable: true, get: function () { return events_js_1.isRuntimeEvent; } });
Object.defineProperty(exports, "isRuntimeEventContext", { enumerable: true, get: function () { return events_js_1.isRuntimeEventContext; } });
Object.defineProperty(exports, "isRuntimeEventOfType", { enumerable: true, get: function () { return events_js_1.isRuntimeEventOfType; } });
var run_js_1 = require("./run.js");
Object.defineProperty(exports, "Run", { enumerable: true, get: function () { return run_js_1.Run; } });
var runtime_js_2 = require("./runtime.js");
Object.defineProperty(exports, "phases", { enumerable: true, get: function () { return runtime_js_2.phases; } });
/**
 * Create new instance of the Postman Runtime
 *
 * @example
 * ```ts
 * const runtime = new Runtime();
 * const run = runtime.execute({
 *   type: 'collection',
 *   children: [...]
 * });
 *
 * for await (const event of run) {
 *   console.log(event);
 * }
 * ```
 */
class Runtime extends runtime_js_1.Runtime {
    constructor(config) {
        // Allow partial overrides of built-in platform at namespace level
        super({ ...config, platform: { ...config?.platform, ...platform_node_js_1.platform } });
    }
}
exports.Runtime = Runtime;
