// @postman/runtime/dist/lib/objects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPlainObject = isPlainObject;
exports.isEmpty = isEmpty;
exports.isShallowEqual = isShallowEqual;
exports.pick = pick;
exports.omit = omit;
exports.defaulted = defaulted;
exports.get = get;
exports.set = set;
const POJO_PROTOTYPE = Object.getPrototypeOf(new Object());
//
// #region Checks
//
/**
 * Check if given value is a Plain Ol' JavaScript Object (POJO)
 * created with `{}`, `new Object({})`, or `Object.create(null, {})`
 */
function isPlainObject(value) {
    if (value == null || typeof value !== 'object')
        return false;
    // Object.create(null, ...) or has `Object` prototype
    const proto = Object.getPrototypeOf(value);
    return proto === null || proto === POJO_PROTOTYPE;
}
/**
 * Check if the given object is empty
 */
function isEmpty(object) {
    return (typeof object !== 'object' || object == null || !Object.keys(object).length);
}
/**
 * Check if first level of objects are equal
 */
function isShallowEqual(a, b) {
    if (a === b)
        return true;
    if (a == null || b == null)
        return false;
    if (typeof a !== 'object' || typeof b !== 'object')
        return false;
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length)
        return false;
    for (const key of keysA) {
        if (!(key in b) || a[key] !== b[key]) {
            return false;
        }
    }
    return true;
}
//
// #region Operations
//
/**
 * Pick specific properties from an object
 */
function pick(object, keys) {
    const result = {};
    for (const key of keys) {
        if (key in object) {
            result[key] = object[key];
        }
    }
    return result;
}
/**
 * Omit specific properties from as object
 */
function omit(object, keys) {
    const result = {};
    for (const key in object) {
        // (types are a little more difficult due to !TKey)
        if (!keys.includes(key)) {
            result[key] = object[key];
        }
    }
    return result;
}
/**
 * Assign default values to object if undefined
 */
function defaulted(object, defaults) {
    const defaulted = Object.assign({}, object);
    for (const key in defaults) {
        if (!(key in object)) {
            defaulted[key] = defaults[key];
        }
    }
    return defaulted;
}
/**
 * Get the value at the given path from the given object
 *
 * Path:
 *
 * - dotted: "a.b.c"
 * - array: ["a", "b", "c"]
 *
 * @example
 * ```ts
 * const object = {
 *   a: 1,
 *   b: { c: { d: 4 } },
 *   'e.f': 6
 * };
 *
 * get(object, 'b.c.d') // -> 4
 * get(object, ['b', 'c', 'd']) // -> 4
 * get(object, 'b.c.z', 42) // -> 42
 * get(object, 'e.f') // -> undefined
 * get(object, ['e.f']) // -> 6
 * ```
 */
function get(object, path, defaultValue) {
    path = Array.isArray(path) ? path : path.split('.');
    let current = object;
    for (let i = 0; i < path.length; i += 1) {
        if (current == null) {
            return defaultValue;
        }
        current = current[path[i]];
    }
    return current ?? defaultValue;
}
/**
 * Set the value at the given path for the given object
 *
 * Path:
 *
 * - dotted: "a.b.c"
 * - array: ["a", "b", "c"]
 *
 * @example
 * ```ts
 * const object = {
 *   a: 1,
 *   b: { c: { d: 4 } },
 *   'e.f': 6
 * };
 *
 * set(object, 'b.c.d', 5)
 * set(object, ['x', 'y', 'z'], 42)
 * ```
 */
function set(object, path, value) {
    path = Array.isArray(path) ? path : path.split('.');
    let current = object;
    for (let i = 0; i < path.length - 1; i += 1) {
        const part = path[i];
        if (typeof current[part] !== 'object' || current[part] === null) {
            current[part] = {};
        }
        current = current[part];
    }
    current[path[path.length - 1]] = value;
    return object;
}
