// @postman-app/app-updater/src/index.js
const WINDOWS = 'Windows_NT';
const DARWIN = 'Darwin';
const LINUX = 'Linux';

/**
 * @method initializeUpdater
 * @description Provides the updater class Instance based on the OS.
 * It uses Node.os.type() to detect the os.
 * https://nodejs.org/dist/latest-v8.x/docs/api/os.html#os_os_type
 *
 * @returns {?Object} WindowsAutoUpdate || DarwinUpdater || LinuxUpdater || null
 */
function init(options = {}) {
  const { adapter } = options;
  const os = require('os').type();
  const updaterOptions = { adapter, autoUpdater: adapter.getAutoUpdater() };

  switch (os) {
    case WINDOWS: {
      const WindowsUpdater = require('./WindowsUpdater');
      return new WindowsUpdater(updaterOptions);
    }
    case DARWIN: {
      const DarwinUpdater = require('./DarwinUpdater');
      return new DarwinUpdater(updaterOptions);
    }
    case LINUX: {
      const LinuxUpdater = require('./LinuxUpdater');
      const LinuxAutoUpdater = require('./autoUpdater/LinuxAutoUpdater');
      return new LinuxUpdater(
        Object.assign(updaterOptions, {
          autoUpdater: new LinuxAutoUpdater({
            quitApp: adapter.quitApp,
            logger: adapter.getLogger(),
            appInstallationPath: adapter.getAppInstallationPath(),
            appName: adapter.getAppName(),
            originalFs: adapter.getOriginalFs(),
          }),
        })
      );
    }
    default:
      return null;
  }
}

module.exports = { init };
