// @postman/runtime.models/dist/extensions/__tests__/auth.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const auth_1 = require("../auth");
const basic = { type: 'basic', basic: [] };
const oauth2 = {
    type: 'oauth2',
    oauth2: [],
};
const multiAuth = {
    default: '1',
    auth: [
        {
            id: '1',
            name: 'Basic Auth',
            ...basic,
        },
    ],
};
(0, vitest_1.test)('should constrain to certain auth types', () => {
    const Specific = auth_1.Auth.constraints({
        supportedAuthTypes: ['basic', 'bearer'],
    });
    (0, vitest_1.expect)(Specific.parse(basic)).toEqual(basic);
    (0, vitest_1.expect)(Specific.validate(oauth2)).toMatchInlineSnapshot(`
		{
		  "issues": [
		    {
		      "code": "invalid_value",
		      "message": "Unsupported auth type "oauth2"",
		      "path": [
		        "type",
		      ],
		      "values": [
		        "basic",
		        "bearer",
		      ],
		    },
		  ],
		}
	`);
});
(0, vitest_1.test)('should not allow multi-auth by default', () => {
    const Specific = auth_1.Auth.constraints({
        supportedAuthTypes: ['basic', 'bearer'],
        isMultiAuthSupported: false,
    });
    (0, vitest_1.expect)(Specific.validate(multiAuth)).toMatchInlineSnapshot(`
		{
		  "issues": [
		    {
		      "code": "invalid_value",
		      "message": "Multi-auth is not supported",
		      "path": [
		        "auth",
		      ],
		      "values": [],
		    },
		  ],
		}
	`);
});
(0, vitest_1.test)('should allow multi-auth with supported auth types', () => {
    const Specific = auth_1.Auth.constraints({
        supportedAuthTypes: ['basic'],
        isMultiAuthSupported: true,
    });
    (0, vitest_1.expect)(Specific.parse(multiAuth)).toEqual(multiAuth);
});
(0, vitest_1.test)('should not allow multi-auth with unsupported auth types', () => {
    const Specific = auth_1.Auth.constraints({
        supportedAuthTypes: ['basic'],
        isMultiAuthSupported: true,
    });
    (0, vitest_1.expect)(Specific.validate({
        default: '1',
        auth: [
            {
                id: '1',
                name: 'OAuth2 Auth',
                type: 'oauth2',
                oauth2: [],
            },
        ],
    })).toMatchInlineSnapshot(`
		{
		  "issues": [
		    {
		      "code": "invalid_value",
		      "message": "Unsupported auth type "oauth2" in multi-auth",
		      "path": [
		        "auth",
		      ],
		      "values": [
		        "basic",
		      ],
		    },
		  ],
		}
	`);
});
