// @postman/runtime/dist/plugins/auth/oauth2/oauth2-auth-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.oauth2Auth = exports.OAuth2AuthConfig = void 0;
const runtime_models_1 = require("@postman/runtime.models");
const z = __importStar(require("zod/v4-mini"));
const key_values_1 = require("../../../lib/key-values");
const auth_1 = require("../auth");
exports.OAuth2AuthConfig = z.object({
    accessToken: z.string(),
    addTokenTo: z.pipe(z.optional(z.string()), z.transform((value) => {
        return value?.toLowerCase() === 'queryparams' ? 'queryParams' : 'header';
    })),
    tokenType: z.pipe(z.optional(z.string()), z.transform((value) => {
        return value?.toLowerCase() === 'mac' ? 'mac' : 'bearer';
    })),
    headerPrefix: z.pipe(z.optional(z.string()), z.transform((value = 'Bearer') => {
        const trimmed = value.trim();
        if (!trimmed)
            return '';
        return `${trimmed} `;
    })),
    // TODO id, refreshOAuth2Token
});
exports.oauth2Auth = {
    name: 'oauth2-auth',
    resolved: {
        'http:send-request'(event, run) {
            if (!runtime_models_1.HTTPRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'oauth2');
            if (!auth)
                return;
            const result = exports.OAuth2AuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.oauth2 ?? []));
            if (!result.success)
                return;
            const config = result.data;
            const { accessToken, addTokenTo = 'header', tokenType, headerPrefix, } = config;
            // TODO Add support for HMAC
            if (tokenType === 'mac')
                return;
            run.patch(event, (payload) => {
                (0, key_values_1.removeKeyValue)(payload.headers, 'authorization', {
                    caseInsensitive: true,
                });
                (0, key_values_1.removeKeyValue)(payload.queryParams, 'access_token', {
                    caseInsensitive: true,
                });
                if (addTokenTo === 'queryParams') {
                    payload.queryParams.push({
                        key: 'access_token',
                        value: accessToken,
                    });
                }
                else {
                    payload.headers.push({
                        key: 'authorization',
                        value: `${headerPrefix}${accessToken}`,
                        system: true,
                    });
                }
            });
        },
        'mcp:request-connect'(event, run) {
            if (!runtime_models_1.MCPRequest.is(run.item) || event.payload.transport !== 'sse')
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'oauth2');
            if (!auth)
                return;
            const result = exports.OAuth2AuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.oauth2 ?? []));
            if (!result.success)
                return;
            const config = result.data;
            const { accessToken, tokenType, headerPrefix } = config;
            // TODO Add support for HMAC
            if (tokenType === 'mac')
                return;
            run.patch(event, (payload) => {
                if (payload.transport === 'sse') {
                    payload.headers ??= [];
                    (0, key_values_1.upsertKeyValue)(payload.headers, {
                        key: 'authorization',
                        value: `${headerPrefix}${accessToken}`,
                    });
                }
            });
        },
        'grpc:invoke'(event, run) {
            if (!runtime_models_1.GRPCRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'oauth2');
            if (!auth)
                return;
            const result = exports.OAuth2AuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.oauth2 ?? []));
            if (!result.success)
                return;
            const config = result.data;
            const { accessToken, tokenType, headerPrefix } = config;
            // TODO Add support for HMAC
            if (tokenType === 'mac')
                return;
            run.patch(event, (payload) => {
                payload.metadata ??= [];
                (0, key_values_1.upsertKeyValue)(payload.metadata, { key: 'Authorization', value: `${headerPrefix}${accessToken}` }, { caseInsensitive: true });
            });
        },
    },
};
