// @postman/runtime.dynamic-variables/dist/default-generators.js
"use strict";
// Adopted from: https://github.com/postmanlabs/postman-collection/blob/v4.3.0/lib/superstring/dynamic-variables.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const en_1 = __importDefault(require("@faker-js/faker/locale/en"));
const uuid_1 = require("uuid");
// Generators for some default dynamic variables
const defaultDynamicGenerators = {
    $guid: {
        description: 'A v4 style guid',
        generator: function () {
            return (0, uuid_1.v4)();
        },
    },
    $timestamp: {
        description: 'The current timestamp',
        generator: function () {
            return Math.round(Date.now() / 1000);
        },
    },
    $isoTimestamp: {
        description: 'The current ISO timestamp at zero UTC',
        generator: function () {
            return new Date().toISOString();
        },
    },
    $randomInt: {
        description: 'A random integer between 0 and 1000',
        generator: function () {
            return ~~(Math.random() * (1000 + 1));
        },
    },
    // faker.phone.phoneNumber returns phone number with or without
    // extension randomly. this only returns a phone number without extension.
    $randomPhoneNumber: {
        description: 'A random 10-digit phone number',
        generator: function () {
            return en_1.default.phone.phoneNumberFormat(0);
        },
    },
    // faker.phone.phoneNumber returns phone number with or without
    // extension randomly. this only returns a phone number with extension.
    $randomPhoneNumberExt: {
        description: 'A random phone number with extension (12 digits)',
        generator: function () {
            return (en_1.default.datatype.number({ min: 1, max: 99 }) +
                '-' +
                en_1.default.phone.phoneNumberFormat(0));
        },
    },
    // faker's random.locale only returns 'en'. this returns from a list of
    // random locales
    $randomLocale: {
        description: 'A random two-letter language code (ISO 639-1)',
        generator: function () {
            return en_1.default.random.arrayElement(LOCALES);
        },
    },
    // fakers' random.words returns random number of words between 1, 3.
    // this returns number of words between 2, 5.
    $randomWords: {
        description: 'Some random words',
        generator: function () {
            var words = [], count = en_1.default.datatype.number({ min: 2, max: 5 }), i;
            for (i = 0; i < count; i++) {
                words.push(en_1.default.random.word());
            }
            return words.join(' ');
        },
    },
    // faker's system.filePath retuns nothing. this returns a path for a file.
    $randomFilePath: {
        description: 'A random file path',
        generator: function () {
            return (defaultDynamicGenerators.$randomDirectoryPath.generator() +
                '/' +
                en_1.default.system.fileName());
        },
    },
    // faker's system.directoryPath retuns nothing. this returns a path for
    // a directory.
    $randomDirectoryPath: {
        description: 'A random directory path',
        generator: function () {
            return en_1.default.random.arrayElement(DIRECTORY_PATHS);
        },
    },
    $randomCity: {
        description: 'A random city name',
        generator: en_1.default.address.city,
    },
    $randomStreetName: {
        description: 'A random street name',
        generator: en_1.default.address.streetName,
    },
    $randomStreetAddress: {
        description: 'A random street address (e.g. 1234 Main Street)',
        generator: en_1.default.address.streetAddress,
    },
    $randomCountry: {
        description: 'A random country',
        generator: en_1.default.address.country,
    },
    $randomCountryCode: {
        description: 'A random 2-letter country code (ISO 3166-1 alpha-2)',
        generator: en_1.default.address.countryCode,
    },
    $randomLatitude: {
        description: 'A random latitude coordinate',
        generator: en_1.default.address.latitude,
    },
    $randomLongitude: {
        description: 'A random longitude coordinate',
        generator: en_1.default.address.longitude,
    },
    $randomColor: {
        description: 'A random color',
        generator: en_1.default.commerce.color,
    },
    $randomDepartment: {
        description: 'A random commerce category (e.g. electronics, clothing)',
        generator: en_1.default.commerce.department,
    },
    $randomProductName: {
        description: 'A random product name (e.g. handmade concrete tuna)',
        generator: en_1.default.commerce.productName,
    },
    $randomProductAdjective: {
        description: 'A random product adjective (e.g. tasty, eco-friendly)',
        generator: en_1.default.commerce.productAdjective,
    },
    $randomProductMaterial: {
        description: 'A random product material (e.g. steel, plastic, leather)',
        generator: en_1.default.commerce.productMaterial,
    },
    $randomProduct: {
        description: 'A random product (e.g. shoes, table, chair)',
        generator: en_1.default.commerce.product,
    },
    $randomCompanyName: {
        description: 'A random company name',
        generator: en_1.default.company.companyName,
    },
    $randomCompanySuffix: {
        description: 'A random company suffix (e.g. Inc, LLC, Group)',
        generator: en_1.default.company.companySuffix,
    },
    $randomCatchPhrase: {
        description: 'A random catchphrase',
        generator: en_1.default.company.catchPhrase,
    },
    $randomBs: {
        description: 'A random phrase of business speak',
        generator: en_1.default.company.bs,
    },
    $randomCatchPhraseAdjective: {
        description: 'A random catchphrase adjective',
        generator: en_1.default.company.catchPhraseAdjective,
    },
    $randomCatchPhraseDescriptor: {
        description: 'A random catchphrase descriptor',
        generator: en_1.default.company.catchPhraseDescriptor,
    },
    $randomCatchPhraseNoun: {
        description: 'Randomly generates a catchphrase noun',
        generator: en_1.default.company.catchPhraseNoun,
    },
    $randomBsAdjective: {
        description: 'A random business speak adjective',
        generator: en_1.default.company.bsAdjective,
    },
    $randomBsBuzz: {
        description: 'A random business speak buzzword',
        generator: en_1.default.company.bsBuzz,
    },
    $randomBsNoun: {
        description: 'A random business speak noun',
        generator: en_1.default.company.bsNoun,
    },
    $randomDatabaseColumn: {
        description: 'A random database column name (e.g. updatedAt, token, group)',
        generator: en_1.default.database.column,
    },
    $randomDatabaseType: {
        description: 'A random database type (e.g. tiny int, double, point)',
        generator: en_1.default.database.type,
    },
    $randomDatabaseCollation: {
        description: 'A random database collation (e.g. cp1250_bin)',
        generator: en_1.default.database.collation,
    },
    $randomDatabaseEngine: {
        description: 'A random database engine (e.g. Memory, Archive, InnoDB)',
        generator: en_1.default.database.engine,
    },
    $randomDatePast: {
        description: 'A random past datetime',
        generator: en_1.default.date.past,
    },
    $randomDateFuture: {
        description: 'A random future datetime',
        generator: en_1.default.date.future,
    },
    $randomDateRecent: {
        description: 'A random recent datetime',
        generator: en_1.default.date.recent,
    },
    $randomMonth: {
        description: 'A random month',
        generator: en_1.default.date.month,
    },
    $randomWeekday: {
        description: 'A random weekday',
        generator: en_1.default.date.weekday,
    },
    $randomBankAccount: {
        description: 'A random 8-digit bank account number',
        generator: en_1.default.finance.account,
    },
    $randomBankAccountName: {
        description: 'A random bank account name (e.g. savings account, checking account)',
        generator: en_1.default.finance.accountName,
    },
    $randomCreditCardMask: {
        description: 'A random masked credit card number',
        generator: en_1.default.finance.mask,
    },
    $randomPrice: {
        description: 'A random price between 0.00 and 1000.00',
        generator: en_1.default.finance.amount,
    },
    $randomTransactionType: {
        description: 'A random transaction type (e.g. invoice, payment, deposit)',
        generator: en_1.default.finance.transactionType,
    },
    $randomCurrencyCode: {
        description: 'A random 3-letter currency code (ISO-4217)',
        generator: en_1.default.finance.currencyCode,
    },
    $randomCurrencyName: {
        description: 'A random currency name',
        generator: en_1.default.finance.currencyName,
    },
    $randomCurrencySymbol: {
        description: 'A random currency symbol',
        generator: en_1.default.finance.currencySymbol,
    },
    $randomBitcoin: {
        description: 'A random bitcoin address',
        generator: en_1.default.finance.bitcoinAddress,
    },
    $randomBankAccountIban: {
        description: 'A random 15-31 character IBAN (International Bank Account Number)',
        generator: en_1.default.finance.iban,
    },
    $randomBankAccountBic: {
        description: 'A random BIC (Bank Identifier Code)',
        generator: en_1.default.finance.bic,
    },
    $randomAbbreviation: {
        description: 'A random abbreviation',
        generator: en_1.default.hacker.abbreviation,
    },
    $randomAdjective: {
        description: 'A random adjective',
        generator: en_1.default.hacker.adjective,
    },
    $randomNoun: {
        description: 'A random noun',
        generator: en_1.default.hacker.noun,
    },
    $randomVerb: {
        description: 'A random verb',
        generator: en_1.default.hacker.verb,
    },
    $randomIngverb: {
        description: 'A random verb ending in “-ing”',
        generator: en_1.default.hacker.ingverb,
    },
    $randomPhrase: {
        description: 'A random phrase',
        generator: en_1.default.hacker.phrase,
    },
    $randomAvatarImage: {
        description: 'A random avatar image',
        generator: en_1.default.image.avatar,
    },
    $randomImageUrl: {
        description: 'A URL for a random image',
        generator: en_1.default.image.imageUrl,
    },
    $randomAbstractImage: {
        description: 'A URL for a random abstract image',
        generator: en_1.default.image.abstract,
    },
    $randomAnimalsImage: {
        description: 'A URL for a random animal image',
        generator: en_1.default.image.animals,
    },
    $randomBusinessImage: {
        description: 'A URL for a random stock business image',
        generator: en_1.default.image.business,
    },
    $randomCatsImage: {
        description: 'A URL for a random cat image',
        generator: en_1.default.image.cats,
    },
    $randomCityImage: {
        description: 'A URL for a random city image',
        generator: en_1.default.image.city,
    },
    $randomFoodImage: {
        description: 'A URL for a random food image',
        generator: en_1.default.image.food,
    },
    $randomNightlifeImage: {
        description: 'A URL for a random nightlife image',
        generator: en_1.default.image.nightlife,
    },
    $randomFashionImage: {
        description: 'A URL for a random fashion image',
        generator: en_1.default.image.fashion,
    },
    $randomPeopleImage: {
        description: 'A URL for a random image of a person',
        generator: en_1.default.image.people,
    },
    $randomNatureImage: {
        description: 'A URL for a random nature image',
        generator: en_1.default.image.nature,
    },
    $randomSportsImage: {
        description: 'A URL for a random sports image',
        generator: en_1.default.image.sports,
    },
    $randomTransportImage: {
        description: 'A URL for a random transportation image',
        generator: en_1.default.image.transport,
    },
    $randomImageDataUri: {
        description: 'A random image data URI',
        generator: en_1.default.image.dataUri,
    },
    $randomEmail: {
        description: 'A random email address',
        generator: en_1.default.internet.email,
    },
    $randomExampleEmail: {
        description: 'A random email address from an “example” domain (e.g. ben@example.com)',
        generator: en_1.default.internet.exampleEmail,
    },
    $randomUserName: {
        description: 'A random username',
        generator: en_1.default.internet.userName,
    },
    $randomProtocol: {
        description: 'A random internet protocol',
        generator: en_1.default.internet.protocol,
    },
    $randomUrl: {
        description: 'A random URL',
        generator: en_1.default.internet.url,
    },
    $randomDomainName: {
        description: 'A random domain name (e.g. gracie.biz, trevor.info)',
        generator: en_1.default.internet.domainName,
    },
    $randomDomainSuffix: {
        description: 'A random domain suffix (e.g. .com, .net, .org)',
        generator: en_1.default.internet.domainSuffix,
    },
    $randomDomainWord: {
        description: 'A random unqualified domain name (a name with no dots)',
        generator: en_1.default.internet.domainWord,
    },
    $randomIP: {
        description: 'A random IPv4 address',
        generator: en_1.default.internet.ip,
    },
    $randomIPV6: {
        description: 'A random IPv6 address',
        generator: en_1.default.internet.ipv6,
    },
    $randomUserAgent: {
        description: 'A random user agent',
        generator: en_1.default.internet.userAgent,
    },
    $randomHexColor: {
        description: 'A random hex value',
        generator: en_1.default.internet.color,
    },
    $randomMACAddress: {
        description: 'A random MAC address',
        generator: en_1.default.internet.mac,
    },
    $randomPassword: {
        description: 'A random 15-character alpha-numeric password',
        generator: en_1.default.internet.password,
    },
    $randomLoremWord: {
        description: 'A random word of lorem ipsum text',
        generator: en_1.default.lorem.word,
    },
    $randomLoremWords: {
        description: 'Some random words of lorem ipsum text',
        generator: en_1.default.lorem.words,
    },
    $randomLoremSentence: {
        description: 'A random sentence of lorem ipsum text',
        generator: en_1.default.lorem.sentence,
    },
    $randomLoremSlug: {
        description: 'A random lorem ipsum URL slug',
        generator: en_1.default.lorem.slug,
    },
    $randomLoremSentences: {
        description: 'A random 2-6 sentences of lorem ipsum text',
        generator: en_1.default.lorem.sentences,
    },
    $randomLoremParagraph: {
        description: 'A random paragraph of lorem ipsum text',
        generator: en_1.default.lorem.paragraph,
    },
    $randomLoremParagraphs: {
        description: '3 random paragraphs of lorem ipsum text',
        generator: en_1.default.lorem.paragraphs,
    },
    $randomLoremText: {
        description: 'A random amount of lorem ipsum text',
        generator: en_1.default.lorem.text,
    },
    $randomLoremLines: {
        description: '1-5 random lines of lorem ipsum',
        generator: en_1.default.lorem.lines,
    },
    $randomFirstName: {
        description: 'A random first name',
        generator: en_1.default.name.firstName,
    },
    $randomLastName: {
        description: 'A random last name',
        generator: en_1.default.name.lastName,
    },
    $randomFullName: {
        description: 'A random first and last name',
        generator: en_1.default.name.findName,
    },
    $randomJobTitle: {
        description: 'A random job title (e.g. senior software developer)',
        generator: en_1.default.name.jobTitle,
    },
    $randomNamePrefix: {
        description: 'A random name prefix (e.g. Mr., Mrs., Dr.)',
        generator: en_1.default.name.prefix,
    },
    $randomNameSuffix: {
        description: 'A random name suffix (e.g. Jr., MD, PhD)',
        generator: en_1.default.name.suffix,
    },
    $randomJobDescriptor: {
        description: 'A random job descriptor (e.g., senior, chief, corporate, etc.)',
        generator: en_1.default.name.jobDescriptor,
    },
    $randomJobArea: {
        description: 'A random job area (e.g. branding, functionality, usability)',
        generator: en_1.default.name.jobArea,
    },
    $randomJobType: {
        description: 'A random job type (e.g. supervisor, manager, coordinator, etc.)',
        generator: en_1.default.name.jobType,
    },
    $randomUUID: {
        description: 'A random 36-character UUID',
        generator: en_1.default.datatype.uuid,
    },
    $randomBoolean: {
        description: 'A random boolean value (true/false)',
        generator: en_1.default.datatype.boolean,
    },
    $randomWord: {
        description: 'A random word',
        generator: en_1.default.random.word,
    },
    $randomAlphaNumeric: {
        description: 'A random alpha-numeric character',
        generator: en_1.default.random.alphaNumeric,
    },
    $randomFileName: {
        description: 'A random file name (includes uncommon extensions)',
        generator: en_1.default.system.fileName,
    },
    $randomCommonFileName: {
        description: 'A random file name',
        generator: en_1.default.system.commonFileName,
    },
    $randomMimeType: {
        description: 'A random MIME type',
        generator: en_1.default.system.mimeType,
    },
    $randomCommonFileType: {
        description: 'A random, common file type (e.g., video, text, image, etc.)',
        generator: en_1.default.system.commonFileType,
    },
    $randomCommonFileExt: {
        description: 'A random, common file extension (.doc, .jpg, etc.)',
        generator: en_1.default.system.commonFileExt,
    },
    $randomFileType: {
        description: 'A random file type (includes uncommon file types)',
        generator: en_1.default.system.fileType,
    },
    $randomFileExt: {
        description: 'A random file extension (includes uncommon extensions)',
        generator: en_1.default.system.fileExt,
    },
    $randomSemver: {
        description: 'A random semantic version number',
        generator: en_1.default.system.semver,
    },
};
// Locale list generated from: https://github.com/chromium/chromium/blob/master/ui/base/l10n/l10n_util.cc
const LOCALES = [
    'af',
    'am',
    'an',
    'ar',
    'ast',
    'az',
    'be',
    'bg',
    'bh',
    'bn',
    'br',
    'bs',
    'ca',
    'ceb',
    'ckb',
    'co',
    'cs',
    'cy',
    'da',
    'de',
    'el',
    'en',
    'eo',
    'es',
    'et',
    'eu',
    'fa',
    'fi',
    'fil',
    'fo',
    'fr',
    'fy',
    'ga',
    'gd',
    'gl',
    'gn',
    'gu',
    'ha',
    'haw',
    'he',
    'hi',
    'hmn',
    'hr',
    'ht',
    'hu',
    'hy',
    'ia',
    'id',
    'ig',
    'is',
    'it',
    'ja',
    'jv',
    'ka',
    'kk',
    'km',
    'kn',
    'ko',
    'ku',
    'ky',
    'la',
    'lb',
    'ln',
    'lo',
    'lt',
    'lv',
    'mg',
    'mi',
    'mk',
    'ml',
    'mn',
    'mo',
    'mr',
    'ms',
    'mt',
    'my',
    'nb',
    'ne',
    'nl',
    'nn',
    'no',
    'ny',
    'oc',
    'om',
    'or',
    'pa',
    'pl',
    'ps',
    'pt',
    'qu',
    'rm',
    'ro',
    'ru',
    'sd',
    'sh',
    'si',
    'sk',
    'sl',
    'sm',
    'sn',
    'so',
    'sq',
    'sr',
    'st',
    'su',
    'sv',
    'sw',
    'ta',
    'te',
    'tg',
    'th',
    'ti',
    'tk',
    'to',
    'tr',
    'tt',
    'tw',
    'ug',
    'uk',
    'ur',
    'uz',
    'vi',
    'wa',
    'xh',
    'yi',
    'yo',
    'zh',
    'zu',
];
// Paths for directories
const DIRECTORY_PATHS = [
    '/Applications',
    '/bin',
    '/boot',
    '/boot/defaults',
    '/dev',
    '/etc',
    '/etc/defaults',
    '/etc/mail',
    '/etc/namedb',
    '/etc/periodic',
    '/etc/ppp',
    '/home',
    '/home/user',
    '/home/user/dir',
    '/lib',
    '/Library',
    '/lost+found',
    '/media',
    '/mnt',
    '/net',
    '/Network',
    '/opt',
    '/opt/bin',
    '/opt/include',
    '/opt/lib',
    '/opt/sbin',
    '/opt/share',
    '/private',
    '/private/tmp',
    '/private/var',
    '/proc',
    '/rescue',
    '/root',
    '/sbin',
    '/selinux',
    '/srv',
    '/sys',
    '/System',
    '/tmp',
    '/Users',
    '/usr',
    '/usr/X11R6',
    '/usr/bin',
    '/usr/include',
    '/usr/lib',
    '/usr/libdata',
    '/usr/libexec',
    '/usr/local/bin',
    '/usr/local/src',
    '/usr/obj',
    '/usr/ports',
    '/usr/sbin',
    '/usr/share',
    '/usr/src',
    '/var',
    '/var/log',
    '/var/mail',
    '/var/spool',
    '/var/tmp',
    '/var/yp',
];
exports.default = defaultDynamicGenerators;
