// @postman/runtime/dist/items/websocket/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.websocket = exports.createWebSocketConstructor = exports.OPEN = exports.CONNECTING = exports.CLOSING = exports.CLOSED = void 0;
var create_websocket_1 = require("./create-websocket");
Object.defineProperty(exports, "CLOSED", { enumerable: true, get: function () { return create_websocket_1.CLOSED; } });
Object.defineProperty(exports, "CLOSING", { enumerable: true, get: function () { return create_websocket_1.CLOSING; } });
Object.defineProperty(exports, "CONNECTING", { enumerable: true, get: function () { return create_websocket_1.CONNECTING; } });
Object.defineProperty(exports, "OPEN", { enumerable: true, get: function () { return create_websocket_1.OPEN; } });
Object.defineProperty(exports, "createWebSocketConstructor", { enumerable: true, get: function () { return create_websocket_1.createWebSocketConstructor; } });
var websocket_plugin_1 = require("./websocket-plugin");
Object.defineProperty(exports, "websocket", { enumerable: true, get: function () { return websocket_plugin_1.websocket; } });
