// @postman/runtime/dist/plugins/cookies/cookie-connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CookieConnectionTransform = void 0;
const cookie_store_1 = require("./cookie-store");
/**
 * Transform an event channel's readable stream to handle cookie events,
 * replying on the given writable stream
 *
 * @example
 * ```ts
 * const channel = platform.send({ ... });
 *
 * const readable = channel.readable.pipeThrough(
 * 	new CookieConnectionTransform(run.context.cookies, channel.writable)
 * );
 *
 * // (cookie events are "consumed" and not present in readable)
 * for await (const event of values(readable)) {
 * 	// ...
 * }
 * ```
 */
class CookieConnectionTransform extends TransformStream {
    constructor(jar, writable) {
        const store = (0, cookie_store_1.normalizeCookieStore)(jar.store);
        const writer = writable.getWriter();
        async function findCookies(event) {
            const { domain, path, allowSpecialUseDomain = true } = event.payload;
            let payload;
            try {
                payload = {
                    result: await store.findCookies(domain, path, allowSpecialUseDomain),
                };
            }
            catch (error) {
                payload = { error };
            }
            await writer.ready;
            await writer.write({
                type: 'cookies:find-cookies.response',
                id: event.id,
                payload,
            });
        }
        async function putCookie(event) {
            const { cookie } = event.payload;
            let payload;
            try {
                await store.putCookie(cookie);
                payload = { result: undefined };
            }
            catch (error) {
                payload = { error };
            }
            await writer.ready;
            await writer.write({
                type: 'cookies:put-cookie.response',
                id: event.id,
                payload,
            });
        }
        super({
            transform(event, controller) {
                if (event.type === 'cookies:find-cookies.request') {
                    findCookies(event);
                }
                else if (event.type === 'cookies:put-cookie.request') {
                    putCookie(event);
                }
                else {
                    // All other events passthrough
                    controller.enqueue(event);
                }
            },
        });
    }
}
exports.CookieConnectionTransform = CookieConnectionTransform;
