// @postman/runtime.models/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Variables = exports.UnknownKeyValues = exports.NullableKeyValues = exports.KeyValues = exports.DescriptiveUnknownKeyValues = exports.DescriptiveNullableKeyValues = exports.DescriptiveKeyValues = exports.Schema = exports.Events = exports.Documentation = exports.Auth = exports.extensionsByType = exports.extensions = exports.WebSocketRequest = exports.WebSocketMessage = exports.WebSocketExample = exports.SocketIORequest = exports.SocketIOMessage = exports.SocketIOExample = exports.MQTTRequest = exports.MQTTMessage = exports.MQTTExample = exports.MCPRequest = exports.LLMRequest = exports.HTTPRequest = exports.HTTPExample = exports.GRPCRequest = exports.GRPCExample = exports.GraphQLRequest = exports.GraphQLExample = exports.Folder = exports.Collection = exports.itemsByType = exports.items = exports.model = exports.assertType = exports.metadata = exports.itemType = exports.assertResourceType = void 0;
const auth_js_1 = require("./extensions/auth.js");
Object.defineProperty(exports, "Auth", { enumerable: true, get: function () { return auth_js_1.Auth; } });
const documentation_js_1 = require("./extensions/documentation.js");
Object.defineProperty(exports, "Documentation", { enumerable: true, get: function () { return documentation_js_1.Documentation; } });
const events_js_1 = require("./extensions/events.js");
Object.defineProperty(exports, "Events", { enumerable: true, get: function () { return events_js_1.Events; } });
const schema_js_1 = require("./extensions/schema.js");
Object.defineProperty(exports, "Schema", { enumerable: true, get: function () { return schema_js_1.Schema; } });
const collection_js_1 = require("./item-types/collection.js");
Object.defineProperty(exports, "Collection", { enumerable: true, get: function () { return collection_js_1.Collection; } });
const folder_js_1 = require("./item-types/folder.js");
Object.defineProperty(exports, "Folder", { enumerable: true, get: function () { return folder_js_1.Folder; } });
const graphql_example_js_1 = require("./item-types/graphql-example.js");
Object.defineProperty(exports, "GraphQLExample", { enumerable: true, get: function () { return graphql_example_js_1.GraphQLExample; } });
const graphql_request_js_1 = require("./item-types/graphql-request.js");
Object.defineProperty(exports, "GraphQLRequest", { enumerable: true, get: function () { return graphql_request_js_1.GraphQLRequest; } });
const grpc_example_js_1 = require("./item-types/grpc-example.js");
Object.defineProperty(exports, "GRPCExample", { enumerable: true, get: function () { return grpc_example_js_1.GRPCExample; } });
const grpc_request_js_1 = require("./item-types/grpc-request.js");
Object.defineProperty(exports, "GRPCRequest", { enumerable: true, get: function () { return grpc_request_js_1.GRPCRequest; } });
const http_example_js_1 = require("./item-types/http-example.js");
Object.defineProperty(exports, "HTTPExample", { enumerable: true, get: function () { return http_example_js_1.HTTPExample; } });
const http_request_js_1 = require("./item-types/http-request.js");
Object.defineProperty(exports, "HTTPRequest", { enumerable: true, get: function () { return http_request_js_1.HTTPRequest; } });
const llm_request_js_1 = require("./item-types/llm-request.js");
Object.defineProperty(exports, "LLMRequest", { enumerable: true, get: function () { return llm_request_js_1.LLMRequest; } });
const mcp_request_js_1 = require("./item-types/mcp-request.js");
Object.defineProperty(exports, "MCPRequest", { enumerable: true, get: function () { return mcp_request_js_1.MCPRequest; } });
const mqtt_example_js_1 = require("./item-types/mqtt-example.js");
Object.defineProperty(exports, "MQTTExample", { enumerable: true, get: function () { return mqtt_example_js_1.MQTTExample; } });
const mqtt_message_js_1 = require("./item-types/mqtt-message.js");
Object.defineProperty(exports, "MQTTMessage", { enumerable: true, get: function () { return mqtt_message_js_1.MQTTMessage; } });
const mqtt_request_js_1 = require("./item-types/mqtt-request.js");
Object.defineProperty(exports, "MQTTRequest", { enumerable: true, get: function () { return mqtt_request_js_1.MQTTRequest; } });
const ws_raw_example_js_1 = require("./item-types/ws-raw-example.js");
Object.defineProperty(exports, "WebSocketExample", { enumerable: true, get: function () { return ws_raw_example_js_1.WebSocketExample; } });
const ws_raw_message_js_1 = require("./item-types/ws-raw-message.js");
Object.defineProperty(exports, "WebSocketMessage", { enumerable: true, get: function () { return ws_raw_message_js_1.WebSocketMessage; } });
const ws_raw_request_js_1 = require("./item-types/ws-raw-request.js");
Object.defineProperty(exports, "WebSocketRequest", { enumerable: true, get: function () { return ws_raw_request_js_1.WebSocketRequest; } });
const ws_socketio_example_js_1 = require("./item-types/ws-socketio-example.js");
Object.defineProperty(exports, "SocketIOExample", { enumerable: true, get: function () { return ws_socketio_example_js_1.SocketIOExample; } });
const ws_socketio_message_js_1 = require("./item-types/ws-socketio-message.js");
Object.defineProperty(exports, "SocketIOMessage", { enumerable: true, get: function () { return ws_socketio_message_js_1.SocketIOMessage; } });
const ws_socketio_request_js_1 = require("./item-types/ws-socketio-request.js");
Object.defineProperty(exports, "SocketIORequest", { enumerable: true, get: function () { return ws_socketio_request_js_1.SocketIORequest; } });
var item_type_js_1 = require("./item-type.js");
Object.defineProperty(exports, "assertResourceType", { enumerable: true, get: function () { return item_type_js_1.assertResourceType; } });
Object.defineProperty(exports, "itemType", { enumerable: true, get: function () { return item_type_js_1.itemType; } });
var metadata_js_1 = require("./metadata.js");
Object.defineProperty(exports, "metadata", { enumerable: true, get: function () { return metadata_js_1.metadata; } });
var model_js_1 = require("./model.js");
Object.defineProperty(exports, "assertType", { enumerable: true, get: function () { return model_js_1.assertType; } });
Object.defineProperty(exports, "model", { enumerable: true, get: function () { return model_js_1.model; } });
// Items
exports.items = [
    collection_js_1.Collection,
    folder_js_1.Folder,
    graphql_example_js_1.GraphQLExample,
    graphql_request_js_1.GraphQLRequest,
    grpc_example_js_1.GRPCExample,
    grpc_request_js_1.GRPCRequest,
    http_example_js_1.HTTPExample,
    http_request_js_1.HTTPRequest,
    llm_request_js_1.LLMRequest,
    mcp_request_js_1.MCPRequest,
    mqtt_example_js_1.MQTTExample,
    mqtt_message_js_1.MQTTMessage,
    mqtt_request_js_1.MQTTRequest,
    ws_raw_example_js_1.WebSocketExample,
    ws_raw_message_js_1.WebSocketMessage,
    ws_raw_request_js_1.WebSocketRequest,
    ws_socketio_example_js_1.SocketIOExample,
    ws_socketio_message_js_1.SocketIOMessage,
    ws_socketio_request_js_1.SocketIORequest,
];
exports.itemsByType = {
    collection: collection_js_1.Collection,
    folder: folder_js_1.Folder,
    'graphql-example': graphql_example_js_1.GraphQLExample,
    'graphql-request': graphql_request_js_1.GraphQLRequest,
    'grpc-example': grpc_example_js_1.GRPCExample,
    'grpc-request': grpc_request_js_1.GRPCRequest,
    'http-example': http_example_js_1.HTTPExample,
    'http-request': http_request_js_1.HTTPRequest,
    'llm-request': llm_request_js_1.LLMRequest,
    'mcp-request': mcp_request_js_1.MCPRequest,
    'mqtt-example': mqtt_example_js_1.MQTTExample,
    'mqtt-message': mqtt_message_js_1.MQTTMessage,
    'mqtt-request': mqtt_request_js_1.MQTTRequest,
    'ws-raw-example': ws_raw_example_js_1.WebSocketExample,
    'ws-raw-message': ws_raw_message_js_1.WebSocketMessage,
    'ws-raw-request': ws_raw_request_js_1.WebSocketRequest,
    'ws-socketio-example': ws_socketio_example_js_1.SocketIOExample,
    'ws-socketio-message': ws_socketio_message_js_1.SocketIOMessage,
    'ws-socketio-request': ws_socketio_request_js_1.SocketIORequest,
};
// Extensions
exports.extensions = [auth_js_1.Auth, documentation_js_1.Documentation, events_js_1.Events, schema_js_1.Schema];
exports.extensionsByType = {
    auth: auth_js_1.Auth,
    documentation: documentation_js_1.Documentation,
    events: events_js_1.Events,
    schema: schema_js_1.Schema,
};
// Common
var key_value_js_1 = require("./common/key-value.js");
Object.defineProperty(exports, "DescriptiveKeyValues", { enumerable: true, get: function () { return key_value_js_1.DescriptiveKeyValues; } });
Object.defineProperty(exports, "DescriptiveNullableKeyValues", { enumerable: true, get: function () { return key_value_js_1.DescriptiveNullableKeyValues; } });
Object.defineProperty(exports, "DescriptiveUnknownKeyValues", { enumerable: true, get: function () { return key_value_js_1.DescriptiveUnknownKeyValues; } });
Object.defineProperty(exports, "KeyValues", { enumerable: true, get: function () { return key_value_js_1.KeyValues; } });
Object.defineProperty(exports, "NullableKeyValues", { enumerable: true, get: function () { return key_value_js_1.NullableKeyValues; } });
Object.defineProperty(exports, "UnknownKeyValues", { enumerable: true, get: function () { return key_value_js_1.UnknownKeyValues; } });
var variables_js_1 = require("./common/variables.js");
Object.defineProperty(exports, "Variables", { enumerable: true, get: function () { return variables_js_1.Variables; } });
