// @postman/runtime.models/dist/item-type.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Id = void 0;
exports.itemType = itemType;
exports.assertResourceType = assertResourceType;
const z = __importStar(require("zod/v4-mini"));
const metadata_1 = require("./metadata");
const model_1 = require("./model");
exports.Id = z.string();
/**
 * Define a new item type
 *
 * @example
 * ```ts
 * const HTTPRequest = itemType({
 *   type: 'http-request',
 *   payload: z.strictObject({
 *     ...
 *   }),
 *   children: {
 *     allowed: [HTTPExample],
 *     // available: allowed, limit
 *   },
 *   extensions: {
 *     optional: {
 *       auth: Auth.constraints({ supportedAuthTypes: [...] }),
 *       documentation: Documentation,
 *       events: Events.constraints({ supportedListeners: [...] }),
 *     }
 *     // available: optional, required
 *   },
 * });
 * ```
 */
function itemType(item) {
    const { allowed: allowedChildTypes = z.never(), limit: limitChildren } = item.children ?? {};
    const extensionConstraints = item.extensions ?? {};
    // Approach:
    // - children: Array<DiscriminatedUnion<ItemType>>
    // - resource.children: Array<Union<...ItemType, Resource>>
    const Children = z.array(Array.isArray(allowedChildTypes) ?
        z.discriminatedUnion('type', allowedChildTypes)
        : allowedChildTypes);
    const ResourceChildren = z.array(Array.isArray(allowedChildTypes) ?
        z.union([
            z.strictObject({
                get type() {
                    return z.literal(allowedChildTypes
                        .map((childType) => {
                        if (isLazy(childType)) {
                            return childType.def.getter().type;
                        }
                        return childType?.type;
                    })
                        .filter((type) => type !== undefined));
                },
                id: exports.Id,
            }),
            ...allowedChildTypes,
        ])
        : z.never());
    const optionalExtensions = Object.fromEntries(Object.entries(extensionConstraints.optional ?? {}).map(([key, value]) => [
        key,
        z.optional(value),
    ]));
    const extensions = extensionConstraints.optional && extensionConstraints.required ?
        z.strictObject({
            ...optionalExtensions,
            ...extensionConstraints.required,
        })
        : extensionConstraints.optional ?
            z.optional(z.strictObject(optionalExtensions))
            : extensionConstraints.required ?
                z.strictObject(extensionConstraints.required)
                : z.optional(z.strictObject({}));
    const ItemType = z.object({
        type: z.literal(item.type),
        id: z.optional(exports.Id),
        title: z.optional(z.string()),
        workspace: z.optional(z.string()),
        createdAt: z.optional(z.string()),
        updatedAt: z.optional(z.string()),
        payload: item.payload,
        children: z.optional(typeof limitChildren === 'number' ?
            Children.check(z.maxLength(limitChildren))
            : Children),
        extensions,
    });
    const ItemModel = (0, model_1.model)(ItemType, {
        type: item.type,
        resource: (0, model_1.model)(z.extend(ItemType, {
            id: exports.Id,
            children: z.optional(typeof limitChildren === 'number' ?
                ResourceChildren.check(z.maxLength(limitChildren))
                : ResourceChildren),
        }), {
            type: item.type,
        }),
    });
    metadata_1.metadata.add(ItemModel, { id: item.type });
    return ItemModel;
}
/**
 * Assert that the given input is a valid value for the given item type's resource schema,
 * throwing an error if invalid
 */
function assertResourceType(itemType, input) {
    const result = itemType.resource.validate(input);
    if (result.issues) {
        throw (0, model_1.invalid)(result);
    }
}
function isLazy(value) {
    return value._zod.def.type === 'lazy';
}
