// @postman/runtime.grpc-client/dist/utilities.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChannelCredentials = getChannelCredentials;
exports.getClientOptions = getClientOptions;
exports.getCallOptions = getCallOptions;
exports.setupProxy = setupProxy;
const grpc = __importStar(require("@postman/grpc-js"));
const tls = __importStar(require("tls"));
// Generates customized ChannelCredentials.
function getChannelCredentials(isTLS, { rejectUnauthorized, secureContext }) {
    if (!isTLS) {
        return grpc.credentials.createInsecure();
    }
    if (!(secureContext instanceof tls.SecureContext)) {
        const { ca, ...options } = secureContext || {};
        secureContext = tls.createSecureContext(options);
        if (ca)
            secureContext.context.addCACert(ca);
    }
    const channelCreds = grpc.credentials.createFromSecureContext(secureContext, {
        rejectUnauthorized: !!rejectUnauthorized,
    });
    return channelCreds;
}
// Generates ClientOptions with support for URL pathnames.
function getClientOptions(parsedURL, settings) {
    return {
        ...settings,
        channelFactoryOverride: (address, credentials, options) => {
            if (parsedURL.pathname && parsedURL.pathname !== '/') {
                return new PathAwareChannel(parsedURL.host, credentials, options, parsedURL.pathname);
            }
            return new grpc.Channel(address, credentials, options);
        },
    };
}
function getCallOptions(options) {
    const callOptions = {};
    if ('deadline' in options) {
        callOptions.deadline = options.deadline;
    }
    else if ('connectionTimeout' in options && options.connectionTimeout) {
        if (typeof options.connectionTimeout !== 'number') {
            throw new TypeError('Invalid connection timeout. Must be a number.');
        }
        callOptions.deadline = Date.now() + options.connectionTimeout;
    }
    return callOptions;
}
/**
 * Setup proxy by setting 'grpc_proxy' environment variable.
 * @param url - The URL of the proxy server.
 */
function setupProxy(url) {
    process.env.grpc_proxy = url;
}
class PathAwareChannel extends grpc.Channel {
    _prefix;
    constructor(target, credentials, options, _prefix) {
        super(target, credentials, options);
        this._prefix = _prefix;
    }
    createCall(method, ...args) {
        return super.createCall(`${this._prefix}${method}`, ...args);
    }
}
