// @postman/runtime/dist/plugins/auth/jwt/jwt-auth-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.jwtAuth = exports.JWTAuthConfig = void 0;
const runtime_models_1 = require("@postman/runtime.models");
const jose_1 = require("jose");
const z = __importStar(require("zod/v4-mini"));
const key_values_1 = require("../../../lib/key-values");
const schemas_1 = require("../../../lib/schemas");
const auth_1 = require("../auth");
const SUPPORTED_SECRET_ALGORITHMS = ['HS256', 'HS384', 'HS512'];
const SUPPORTED_PRIVATE_KEY_ALGORITHMS = [
    'RS256',
    'RS384',
    'RS512',
    'ES256',
    'ES384',
    'ES512',
    'PS256',
    'PS384',
    'PS512',
];
const SUPPORTED_ALGORITHMS = [
    ...SUPPORTED_SECRET_ALGORITHMS,
    ...SUPPORTED_PRIVATE_KEY_ALGORITHMS,
];
// Note: HEADER_ALGORITHM: 'alg' is defined in postman-runtime, but unused
exports.JWTAuthConfig = z.discriminatedUnion('algorithm', [
    z.object({
        algorithm: z.literal(SUPPORTED_SECRET_ALGORITHMS),
        secret: z.string(),
        isSecretBase64Encoded: z.optional(z.boolean()),
        header: z.optional((0, schemas_1.jsonObject)()),
        payload: z.optional((0, schemas_1.jsonObject)()),
        addTokenTo: z.optional(z.literal(['header', 'queryParam'])),
        headerPrefix: z.optional(z.string()),
        queryParamKey: z.optional(z.string()),
    }),
    z.object({
        algorithm: z.literal(SUPPORTED_PRIVATE_KEY_ALGORITHMS),
        privateKey: z.string(),
        header: z.optional((0, schemas_1.jsonObject)()),
        payload: z.optional((0, schemas_1.jsonObject)()),
        addTokenTo: z.optional(z.literal(['header', 'queryParam'])),
        headerPrefix: z.optional(z.string()),
        queryParamKey: z.optional(z.string()),
    }),
]);
/**
 * "jwt" config key-values:
 *
 * - algorithm: string 'HS256', 'HS384', 'HS512', 'RS256', 'RS384', 'RS512', 'ES256', 'ES384', 'ES512', 'PS256', 'PS384', 'PS512'
 * - header: [string] JSON object of header values
 * - payload: [string] JSON object of payload values
 * - privateKey: [string] PEM format private key to use for RS, ES, and PS algorithms
 * - secret: [string] Secret to use for HS algorithms
 * - isSecretBase64Encoded: [boolean = false]
 * - addTokenTo: [string = 'header'] 'header' or 'queryParam'
 * - headerPrefix: [string = 'Bearer']
 * - queryParamKey: [string = 'token']
 */
exports.jwtAuth = {
    name: 'jwt-auth',
    resolved: {
        async 'http:send-request'(event, run) {
            if (!runtime_models_1.HTTPRequest.is(run.item))
                return;
            const auth = (0, auth_1.resolveAuth)(run, 'jwt');
            if (!auth)
                return;
            const result = exports.JWTAuthConfig.safeParse((0, key_values_1.keyValuesToObject)(auth.jwt ?? []));
            if (!result.success)
                return;
            const config = result.data;
            const { header, payload, addTokenTo = 'header', headerPrefix = 'Bearer', queryParamKey = 'token', } = config;
            const token = await signToken(config, { header, payload });
            if (addTokenTo === 'header') {
                run.patch(event, (payload) => {
                    payload.headers.push({
                        key: 'Authorization',
                        value: headerPrefix ? `${headerPrefix} ${token}` : token,
                        system: true,
                    });
                });
            }
            else {
                run.patch(event, (payload) => {
                    payload.queryParams ??= [];
                    payload.queryParams.push({
                        key: queryParamKey,
                        value: token,
                    });
                });
            }
        },
    },
};
async function signToken(config, values) {
    const { algorithm, privateKey, secret, isSecretBase64Encoded = false, } = config;
    const { header, payload } = values;
    const secretData = secret ?
        isSecretBase64Encoded ?
            Uint8Array.from(atob(secret), (value) => value.charCodeAt(0))
            : new TextEncoder().encode(secret)
        : undefined;
    // treat root level "alg" as source of truth
    const headerWithAlg = { ...header, alg: algorithm };
    const signingKey = SUPPORTED_SECRET_ALGORITHMS.includes(algorithm) ?
        secretData
        : await (0, jose_1.importPKCS8)(privateKey, algorithm);
    const token = await new jose_1.SignJWT(payload)
        .setProtectedHeader(headerWithAlg)
        .sign(signingKey);
    return token;
}
