// @postman/runtime/dist/plugins/certificates/certificates-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.certificates = void 0;
const certificates_1 = require("./certificates");
exports.certificates = {
    name: 'certificates',
    resolved: {
        async 'http:send-request'(event, run) {
            await handleCertificates(run, event.payload.url, (certs) => {
                run.patch(event, (payload) => {
                    payload.settings.certificate = certs.certificate;
                    payload.settings.ca = certs.ca;
                });
            });
        },
        async 'grpc:invoke'(event, run) {
            await handleCertificates(run, event.payload.url, (certs) => {
                run.patch(event, (payload) => {
                    payload.settings ??= {};
                    payload.settings.certificate = certs.certificate;
                    payload.settings.ca = certs.ca;
                });
            });
        },
        async 'mcp:request-connect'(event, run) {
            if (event.payload.transport === 'sse') {
                await handleCertificates(run, event.payload.url, (certs) => {
                    run.patch(event, (payload) => {
                        payload.settings = {
                            ...payload.settings,
                            certificate: certs.certificate,
                            ca: certs.ca,
                        };
                    });
                });
            }
        },
        async 'websocket:open'(event, run) {
            await handleCertificates(run, event.payload.url, (certs) => {
                run.patch(event, (payload) => {
                    payload.settings ??= {};
                    payload.settings.certificate = certs.certificate;
                    payload.settings.ca = certs.ca;
                });
            });
        },
    },
};
async function handleCertificates(run, url, patchCallback) {
    if (!run.context.certificates)
        return;
    const { certificate, ca } = await (0, certificates_1.loadCertificate)(run.runtime.platform ?? {}, run.context.certificates, url);
    patchCallback({ certificate, ca });
}
