// @postman/runtime.models/dist/extensions/auth.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auth = exports.MultiAuthSchema = exports.SingleAuthSchema = void 0;
const z = __importStar(require("zod/v4-mini"));
const key_value_js_1 = require("../common/key-value.js");
const metadata_js_1 = require("../metadata.js");
const model_js_1 = require("../model.js");
const NoAuth = {
    type: z.literal('noauth'),
}, AWSV4Auth = {
    type: z.literal('awsv4'),
    awsv4: key_value_js_1.UnknownKeyValues,
}, BasicAuth = {
    type: z.literal('basic'),
    basic: key_value_js_1.UnknownKeyValues,
}, BearerAuth = {
    type: z.literal('bearer'),
    bearer: key_value_js_1.UnknownKeyValues,
}, HawkAuth = {
    type: z.literal('hawk'),
    hawk: key_value_js_1.UnknownKeyValues,
}, OAuth1 = {
    type: z.literal('oauth1'),
    oauth1: key_value_js_1.UnknownKeyValues,
}, OAuth2 = {
    type: z.literal('oauth2'),
    oauth2: key_value_js_1.UnknownKeyValues,
}, APIKeyAuth = {
    type: z.literal('apikey'),
    apikey: key_value_js_1.UnknownKeyValues,
}, EdgeGridAuth = {
    type: z.literal('edgegrid'),
    edgegrid: key_value_js_1.UnknownKeyValues,
}, JWTAuth = {
    type: z.literal('jwt'),
    jwt: key_value_js_1.UnknownKeyValues,
}, DigestAuth = {
    type: z.literal('digest'),
    digest: key_value_js_1.UnknownKeyValues,
}, NTLMAuth = {
    type: z.literal('ntlm'),
    ntlm: key_value_js_1.UnknownKeyValues,
}, ASAPAuth = {
    type: z.literal('asap'),
    asap: key_value_js_1.UnknownKeyValues,
}, InheritAuth = {
    type: z.literal('inherit'),
    inherit: key_value_js_1.UnknownKeyValues,
};
exports.SingleAuthSchema = z.discriminatedUnion('type', [
    z.strictObject(NoAuth),
    z.strictObject(AWSV4Auth),
    z.strictObject(BasicAuth),
    z.strictObject(BearerAuth),
    z.strictObject(HawkAuth),
    z.strictObject(OAuth1),
    z.strictObject(OAuth2),
    z.strictObject(APIKeyAuth),
    z.strictObject(EdgeGridAuth),
    z.strictObject(JWTAuth),
    z.strictObject(DigestAuth),
    z.strictObject(NTLMAuth),
    z.strictObject(ASAPAuth),
    z.strictObject(InheritAuth),
]);
const MultiAuthMeta = {
    id: z.string(),
    name: z.string(),
    description: z.optional(z.string()),
};
exports.MultiAuthSchema = z.strictObject({
    default: z.string(),
    auth: z.array(z.union([
        z.strictObject({
            ...MultiAuthMeta,
            ...NoAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...AWSV4Auth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...BasicAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...BearerAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...HawkAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...OAuth1,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...OAuth2,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...APIKeyAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...EdgeGridAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...JWTAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...DigestAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...NTLMAuth,
        }),
        z.strictObject({
            ...MultiAuthMeta,
            ...ASAPAuth,
        }),
    ])),
});
const AuthSchema = z.union([exports.SingleAuthSchema, exports.MultiAuthSchema]);
function isSingleAuth(auth) {
    return exports.SingleAuthSchema.safeParse(auth).success;
}
exports.Auth = (0, model_js_1.model)(AuthSchema, {
    constraints(options = {}) {
        return (0, model_js_1.model)(AuthSchema.check((context) => {
            if (isSingleAuth(context.value)) {
                if (options.supportedAuthTypes &&
                    !options.supportedAuthTypes.includes(context.value.type)) {
                    context.issues.push({
                        code: 'invalid_value',
                        message: `Unsupported auth type "${context.value.type}"`,
                        path: ['type'],
                        values: options.supportedAuthTypes,
                        input: context.value.type,
                    });
                }
            }
            else {
                if (!options.isMultiAuthSupported) {
                    context.issues.push({
                        code: 'invalid_value',
                        message: 'Multi-auth is not supported',
                        path: ['auth'],
                        input: context.value.auth,
                        values: [],
                    });
                }
                else if (options.supportedAuthTypes) {
                    for (const entry of context.value.auth) {
                        if (!options.supportedAuthTypes.includes(entry.type)) {
                            context.issues.push({
                                code: 'invalid_value',
                                message: `Unsupported auth type "${entry.type}" in multi-auth`,
                                path: ['auth'],
                                input: entry.type,
                                values: options.supportedAuthTypes,
                            });
                        }
                    }
                }
            }
        }));
    },
});
metadata_js_1.metadata.add(exports.Auth, { id: 'auth-extension' });
