// @postman/runtime.models/dist/__tests__/index.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const models = __importStar(require("../index.js"));
(0, vitest_1.test)('Collection', () => {
    (0, vitest_1.expect)(models.Collection.validate({
        type: 'collection',
        payload: {},
        children: [],
        extensions: {
            documentation: { content: '...' },
        },
    }).issues).toBeUndefined();
    // Test with auth extension
    (0, vitest_1.expect)(models.Collection.validate({
        type: 'collection',
        payload: {},
        children: [],
        extensions: {
            documentation: { content: '...' },
            auth: {
                default: '1',
                auth: [
                    {
                        id: '1',
                        name: 'Basic Auth',
                        type: 'basic',
                        basic: [],
                    },
                ],
            },
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('Folder', () => {
    (0, vitest_1.expect)(models.Folder.validate({
        type: 'folder',
        payload: {},
        children: [{ type: 'folder', payload: {} }],
    }).issues).toBeUndefined();
    // Test with auth extension
    (0, vitest_1.expect)(models.Folder.validate({
        type: 'folder',
        payload: {},
        children: [{ type: 'folder', payload: {} }],
        extensions: {
            auth: {
                default: '1',
                auth: [
                    {
                        id: '1',
                        name: 'Basic Auth',
                        type: 'basic',
                        basic: [],
                    },
                ],
            },
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('GraphQLRequest', () => {
    (0, vitest_1.expect)(models.GraphQLRequest.validate({
        type: 'graphql-request',
        payload: {
            url: '...',
            query: '...',
        },
    }).issues).toBeUndefined();
    // Test with auth extension
    (0, vitest_1.expect)(models.GraphQLRequest.validate({
        type: 'graphql-request',
        payload: {
            url: '...',
            query: '...',
        },
        extensions: {
            auth: {
                type: 'inherit',
                inherit: [
                    {
                        key: 'authId',
                        value: '1',
                    },
                ],
            },
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('GraphQLExample', () => {
    (0, vitest_1.expect)(models.GraphQLExample.validate({
        type: 'graphql-example',
        payload: {
            url: '...',
            query: '...',
        },
    }).issues).toBeUndefined();
    // Test with auth extension
    (0, vitest_1.expect)(models.GraphQLExample.validate({
        type: 'graphql-example',
        payload: {
            url: '...',
            query: '...',
        },
        extensions: {
            auth: {
                type: 'bearer',
                bearer: [
                    {
                        key: 'token',
                        value: '1',
                    },
                ],
            },
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('GRPCRequest', () => {
    (0, vitest_1.expect)(models.GRPCRequest.validate({
        type: 'grpc-request',
        payload: {
            url: '...',
            methodPath: '...',
            message: { content: '...' },
            metadata: [],
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('GRPCExample', () => {
    (0, vitest_1.expect)(models.GRPCExample.validate({
        type: 'grpc-example',
        payload: {
            url: '...',
            methodPath: '...',
            message: { content: '...' },
            metadata: [],
            response: {
                code: 0,
                metadata: [],
                trailers: [],
                messages: [],
            },
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('HTTPRequest', () => {
    (0, vitest_1.expect)(models.HTTPRequest.validate({
        type: 'http-request',
        payload: {
            url: '...',
            method: 'GET',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('HTTPExample', () => {
    (0, vitest_1.expect)(models.HTTPExample.validate({
        type: 'http-example',
        payload: {
            url: '...',
            method: 'GET',
            headers: [],
            queryParams: [],
            response: {
                statusCode: 200,
                statusText: 'OK',
                body: {
                    type: 'text',
                    content: '...',
                },
            },
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('LLMRequest', () => {
    (0, vitest_1.expect)(models.LLMRequest.validate({
        type: 'llm-request',
        payload: {
            url: '...',
            config: {
                model: '...',
                provider: '...',
            },
            settings: {},
            userPrompts: [],
            systemPrompts: [],
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('MCPRequest', () => {
    (0, vitest_1.expect)(models.MCPRequest.validate({
        type: 'mcp-request',
        payload: {
            transport: 'sse',
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('MQTTRequest', () => {
    (0, vitest_1.expect)(models.MQTTRequest.validate({
        type: 'mqtt-request',
        payload: {
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('MQTTMessage', () => {
    (0, vitest_1.expect)(models.MQTTMessage.validate({
        type: 'mqtt-message',
        payload: {
            payload: '...',
            topic: '...',
            type: 'text',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('MQTTExample', () => {
    (0, vitest_1.expect)(models.MQTTExample.validate({
        type: 'mqtt-example',
        payload: {
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('WebSocketRequest', () => {
    (0, vitest_1.expect)(models.WebSocketRequest.validate({
        type: 'ws-raw-request',
        payload: {
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('WebSocketMessage', () => {
    (0, vitest_1.expect)(models.WebSocketMessage.validate({
        type: 'ws-raw-message',
        payload: {
            type: 'text',
            content: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('WebSocketExample', () => {
    (0, vitest_1.expect)(models.WebSocketExample.validate({
        type: 'ws-raw-example',
        payload: {
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('SocketIORequest', () => {
    (0, vitest_1.expect)(models.SocketIORequest.validate({
        type: 'ws-socketio-request',
        payload: {
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('SocketIOMessage', () => {
    (0, vitest_1.expect)(models.SocketIOMessage.validate({
        type: 'ws-socketio-message',
        payload: {
            eventName: '...',
            args: [
                {
                    type: 'text',
                    content: '...',
                },
            ],
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('SocketIOExample', () => {
    (0, vitest_1.expect)(models.SocketIOExample.validate({
        type: 'ws-socketio-example',
        payload: {
            url: '...',
        },
    }).issues).toBeUndefined();
});
(0, vitest_1.test)('items', () => {
    (0, vitest_1.expect)(models.items).toEqual([
        models.Collection,
        models.Folder,
        models.GraphQLExample,
        models.GraphQLRequest,
        models.GRPCExample,
        models.GRPCRequest,
        models.HTTPExample,
        models.HTTPRequest,
        models.LLMRequest,
        models.MCPRequest,
        models.MQTTExample,
        models.MQTTMessage,
        models.MQTTRequest,
        models.WebSocketExample,
        models.WebSocketMessage,
        models.WebSocketRequest,
        models.SocketIOExample,
        models.SocketIOMessage,
        models.SocketIORequest,
    ]);
});
(0, vitest_1.test)('itemsByType', () => {
    (0, vitest_1.expect)(models.itemsByType).toEqual({
        collection: models.Collection,
        folder: models.Folder,
        'graphql-example': models.GraphQLExample,
        'graphql-request': models.GraphQLRequest,
        'grpc-example': models.GRPCExample,
        'grpc-request': models.GRPCRequest,
        'http-example': models.HTTPExample,
        'http-request': models.HTTPRequest,
        'llm-request': models.LLMRequest,
        'mcp-request': models.MCPRequest,
        'mqtt-example': models.MQTTExample,
        'mqtt-message': models.MQTTMessage,
        'mqtt-request': models.MQTTRequest,
        'ws-raw-example': models.WebSocketExample,
        'ws-raw-message': models.WebSocketMessage,
        'ws-raw-request': models.WebSocketRequest,
        'ws-socketio-example': models.SocketIOExample,
        'ws-socketio-message': models.SocketIOMessage,
        'ws-socketio-request': models.SocketIORequest,
    });
});
(0, vitest_1.test)('extensions', () => {
    (0, vitest_1.expect)(models.extensions).toEqual([
        models.Auth,
        models.Documentation,
        models.Events,
        models.Schema,
    ]);
});
(0, vitest_1.test)('extensionsByType', () => {
    (0, vitest_1.expect)(models.extensionsByType).toEqual({
        auth: models.Auth,
        documentation: models.Documentation,
        events: models.Events,
        schema: models.Schema,
    });
});
