// @postman/runtime.models/dist/__tests__/item-type.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const z = __importStar(require("zod/v4-mini"));
const auth_js_1 = require("../extensions/auth.js");
const documentation_js_1 = require("../extensions/documentation.js");
const item_type_js_1 = require("../item-type.js");
const folder_js_1 = require("../item-types/folder.js");
const ExampleMessage = (0, item_type_js_1.itemType)({
    type: 'example-message',
    payload: z.strictObject({
        content: z.string(),
    }),
    // limit-children: 0
    children: {
        limit: 0,
    },
});
const ExampleRequest = (0, item_type_js_1.itemType)({
    type: 'example-request',
    payload: z.strictObject({
        url: z.string(),
    }),
    // allow-child-types: ["example-message"]
    children: {
        allowed: [ExampleMessage],
    },
    extensions: {
        optional: {
            auth: auth_js_1.Auth.constraints({ supportedAuthTypes: ['basic'] }),
        },
    },
});
const Combined = (0, item_type_js_1.itemType)({
    type: 'combined',
    payload: z.strictObject({
        value: z.number(),
    }),
    extensions: {
        optional: {
            auth: auth_js_1.Auth,
        },
        required: {
            documentation: documentation_js_1.Documentation,
        },
    },
});
const Collection = (0, item_type_js_1.itemType)({
    type: 'collection',
    payload: z.strictObject({
        variables: z.array(z.strictObject({
            key: z.string(),
            value: z.string(),
            disabled: z.optional(z.boolean()),
        })),
    }),
    children: {
        allowed: [ExampleRequest, Combined],
    },
    // allow-extensions: ["documentation"]
    // require-extensions: ["documentation"]
    extensions: {
        required: {
            documentation: documentation_js_1.Documentation,
        },
    },
});
(0, vitest_1.test)('should handle valid item', () => {
    (0, vitest_1.expect)(ExampleRequest.validate({
        type: 'example-request',
        payload: { url: 'localhost:3000' },
    })).toMatchInlineSnapshot(`
		{
		  "value": {
		    "payload": {
		      "url": "localhost:3000",
		    },
		    "type": "example-request",
		  },
		}
	`);
    (0, vitest_1.expect)(ExampleRequest.validate({
        type: 'example-request',
        payload: { url: 'localhost:3000' },
        children: [{ type: 'example-message', payload: { content: 'Howdy!' } }],
        extensions: { auth: { type: 'basic', basic: [] } },
    })).toMatchInlineSnapshot(`
		{
		  "value": {
		    "children": [
		      {
		        "payload": {
		          "content": "Howdy!",
		        },
		        "type": "example-message",
		      },
		    ],
		    "extensions": {
		      "auth": {
		        "basic": [],
		        "type": "basic",
		      },
		    },
		    "payload": {
		      "url": "localhost:3000",
		    },
		    "type": "example-request",
		  },
		}
	`);
});
(0, vitest_1.describe)('constraints', () => {
    // allow-extensions
    (0, vitest_1.test)('should support allow-extensions', () => {
        // positive
        (0, vitest_1.expect)(ExampleRequest.validate({
            type: 'example-request',
            payload: { url: '...' },
            extensions: {
                auth: { type: 'basic', basic: [] },
            },
        })).toMatchInlineSnapshot(`
			{
			  "value": {
			    "extensions": {
			      "auth": {
			        "basic": [],
			        "type": "basic",
			      },
			    },
			    "payload": {
			      "url": "...",
			    },
			    "type": "example-request",
			  },
			}
		`);
        // negative
        (0, vitest_1.expect)(ExampleRequest.validate({
            type: 'example-request',
            payload: { url: '...' },
            extensions: {
                auth: { type: 'basic', basic: [] },
                notallowed: {},
            },
        })).toMatchInlineSnapshot(`
			{
			  "issues": [
			    {
			      "code": "unrecognized_keys",
			      "keys": [
			        "notallowed",
			      ],
			      "message": "Invalid input",
			      "path": [
			        "extensions",
			      ],
			    },
			  ],
			}
		`);
    });
    // require-extensions
    (0, vitest_1.test)('should support require-extensions', () => {
        // positive
        (0, vitest_1.expect)(Collection.validate({
            type: 'collection',
            payload: { variables: [] },
            extensions: {
                documentation: { content: '...' },
            },
        })).toMatchInlineSnapshot(`
			{
			  "value": {
			    "extensions": {
			      "documentation": {
			        "content": "...",
			      },
			    },
			    "payload": {
			      "variables": [],
			    },
			    "type": "collection",
			  },
			}
		`);
        // negative
        (0, vitest_1.expect)(Collection.validate({
            type: 'collection',
            payload: { variables: [] },
            extensions: {
                docs: { content: '...' },
            },
        })).toMatchInlineSnapshot(`
			{
			  "issues": [
			    {
			      "code": "invalid_type",
			      "expected": "object",
			      "message": "Invalid input",
			      "path": [
			        "extensions",
			        "documentation",
			      ],
			    },
			    {
			      "code": "unrecognized_keys",
			      "keys": [
			        "docs",
			      ],
			      "message": "Invalid input",
			      "path": [
			        "extensions",
			      ],
			    },
			  ],
			}
		`);
    });
    // allow-extensions + require-extensions
    (0, vitest_1.test)('should support allow-extensions + require-extensions', () => {
        // positive
        (0, vitest_1.expect)(Combined.validate({
            type: 'combined',
            payload: {
                value: 42,
            },
            extensions: {
                documentation: { content: '...' },
            },
        })).toMatchInlineSnapshot(`
			{
			  "value": {
			    "extensions": {
			      "documentation": {
			        "content": "...",
			      },
			    },
			    "payload": {
			      "value": 42,
			    },
			    "type": "combined",
			  },
			}
		`);
        // negative
        (0, vitest_1.expect)(Combined.validate({
            type: 'combined',
            payload: {
                value: 42,
            },
            extensions: {
                auth: { type: 'noauth' },
            },
        })).toMatchInlineSnapshot(`
			{
			  "issues": [
			    {
			      "code": "invalid_type",
			      "expected": "object",
			      "message": "Invalid input",
			      "path": [
			        "extensions",
			        "documentation",
			      ],
			    },
			  ],
			}
		`);
    });
    // allow-child-types
    (0, vitest_1.test)('should support allow-child-types', () => {
        (0, vitest_1.expect)(ExampleRequest.validate({
            type: 'example-request',
            payload: { url: '...' },
            children: [{ type: 'example-message', payload: { content: '...' } }],
        })).toMatchInlineSnapshot(`
			{
			  "value": {
			    "children": [
			      {
			        "payload": {
			          "content": "...",
			        },
			        "type": "example-message",
			      },
			    ],
			    "payload": {
			      "url": "...",
			    },
			    "type": "example-request",
			  },
			}
		`);
        // negative
        (0, vitest_1.expect)(ExampleRequest.validate({
            type: 'example-request',
            payload: { url: '...' },
            children: [
                { type: 'example-message', payload: { content: '...' } },
                { type: 'unknown', payload: { content: '...' } },
            ],
        })).toMatchInlineSnapshot(`
			{
			  "issues": [
			    {
			      "code": "invalid_union",
			      "errors": [],
			      "message": "Invalid input",
			      "note": "No matching discriminator",
			      "path": [
			        "children",
			        1,
			        "type",
			      ],
			    },
			  ],
			}
		`);
    });
    // block-child-types
    // (not supported)
    // limit-children
    (0, vitest_1.test)('should support limit-children', () => {
        // positive
        (0, vitest_1.expect)(ExampleMessage.validate({
            type: 'example-message',
            payload: { content: '...' },
            children: [],
        })).toMatchInlineSnapshot(`
			{
			  "value": {
			    "children": [],
			    "payload": {
			      "content": "...",
			    },
			    "type": "example-message",
			  },
			}
		`);
        // negative
        (0, vitest_1.expect)(ExampleMessage.validate({
            type: 'example-message',
            payload: { content: '...' },
            children: [{ type: 'unknown' }],
        })).toMatchInlineSnapshot(`
			{
			  "issues": [
			    {
			      "code": "invalid_type",
			      "expected": "never",
			      "message": "Invalid input",
			      "path": [
			        "children",
			        0,
			      ],
			    },
			    {
			      "code": "too_big",
			      "inclusive": true,
			      "maximum": 0,
			      "message": "Invalid input",
			      "origin": "array",
			      "path": [
			        "children",
			      ],
			    },
			  ],
			}
		`);
    });
    // limit-children-by-type
    // (not supported)
});
(0, vitest_1.test)('should handle invalid item', () => {
    (0, vitest_1.expect)(ExampleRequest.validate({
        type: 'unknown-request',
        payload: { url: 'localhost:3000' },
    }).issues).toMatchInlineSnapshot(`
		[
		  {
		    "code": "invalid_value",
		    "message": "Invalid input",
		    "path": [
		      "type",
		    ],
		    "values": [
		      "example-request",
		    ],
		  },
		]
	`);
    (0, vitest_1.expect)(ExampleRequest.validate({
        type: 'example-request',
        payload: { url: 42 },
    }).issues).toMatchInlineSnapshot(`
		[
		  {
		    "code": "invalid_type",
		    "expected": "string",
		    "message": "Invalid input",
		    "path": [
		      "payload",
		      "url",
		    ],
		  },
		]
	`);
    (0, vitest_1.expect)(ExampleRequest.validate({
        type: 'example-request',
        payload: { url: 'localhost:3000' },
        children: [{ type: 'example-message', payload: { content: 'Howdy!' } }],
        extensions: {
            auth: { type: 'bearer', bearer: [] },
        },
    }).issues).toMatchInlineSnapshot(`
		[
		  {
		    "code": "invalid_value",
		    "message": "Unsupported auth type "bearer"",
		    "path": [
		      "extensions",
		      "auth",
		      "type",
		    ],
		    "values": [
		      "basic",
		    ],
		  },
		]
	`);
});
(0, vitest_1.describe)('extensions', () => {
    (0, vitest_1.test)('should allow optional extensions', () => {
        const input = {
            type: 'example-request',
            payload: { url: 'localhost:3000' },
            id: '(id)',
            title: '(title)',
            workspace: '(workspace)',
            createdAt: '2025-01-01T00:00:00.000Z',
            updatedAt: '2025-01-01T00:00:00.000Z',
        };
        (0, vitest_1.expect)(ExampleRequest.parse(input)).toEqual(input);
    });
});
(0, vitest_1.describe)('resource', () => {
    (0, vitest_1.test)('should make id required', () => {
        (0, vitest_1.expect)(Collection.resource.validate({
            type: 'collection',
            id: '1',
            payload: { variables: [] },
            extensions: { documentation: { content: '' } },
        }).issues).toBeUndefined();
        (0, vitest_1.expect)(Collection.resource.validate({
            type: 'collection',
            payload: { variables: [] },
            extensions: { documentation: { content: '' } },
        }).issues).toMatchInlineSnapshot(`
			[
			  {
			    "code": "invalid_type",
			    "expected": "string",
			    "message": "Invalid input",
			    "path": [
			      "id",
			    ],
			  },
			]
		`);
    });
    (0, vitest_1.test)('should make type + id for children', () => {
        (0, vitest_1.expect)(Collection.resource.validate({
            type: 'collection',
            id: '1',
            payload: { variables: [] },
            extensions: { documentation: { content: '' } },
            children: [
                { type: 'example-request', id: '2' },
                {
                    type: 'combined',
                    id: '3',
                    payload: { value: 42 },
                    extensions: { documentation: { content: '...' } },
                },
            ],
        }).issues).toBeUndefined();
        (0, vitest_1.expect)(Collection.resource.validate({
            type: 'collection',
            id: '1',
            payload: { variables: [] },
            extensions: { documentation: { content: '' } },
            children: [{ type: 'unknown', id: '3' }],
        }).issues).toMatchInlineSnapshot(`
			[
			  {
			    "code": "invalid_union",
			    "errors": [
			      [
			        {
			          "code": "invalid_value",
			          "message": "Invalid input",
			          "path": [
			            "type",
			          ],
			          "values": [
			            "example-request",
			            "combined",
			          ],
			        },
			      ],
			      [
			        {
			          "code": "invalid_value",
			          "message": "Invalid input",
			          "path": [
			            "type",
			          ],
			          "values": [
			            "example-request",
			          ],
			        },
			        {
			          "code": "invalid_type",
			          "expected": "object",
			          "message": "Invalid input",
			          "path": [
			            "payload",
			          ],
			        },
			      ],
			      [
			        {
			          "code": "invalid_value",
			          "message": "Invalid input",
			          "path": [
			            "type",
			          ],
			          "values": [
			            "combined",
			          ],
			        },
			        {
			          "code": "invalid_type",
			          "expected": "object",
			          "message": "Invalid input",
			          "path": [
			            "payload",
			          ],
			        },
			        {
			          "code": "invalid_type",
			          "expected": "object",
			          "message": "Invalid input",
			          "path": [
			            "extensions",
			          ],
			        },
			      ],
			    ],
			    "message": "Invalid input",
			    "path": [
			      "children",
			      0,
			    ],
			  },
			]
		`);
    });
    (0, vitest_1.test)('should handle Folder resource with nested folder', () => {
        (0, vitest_1.expect)(folder_js_1.Folder.resource.validate({
            type: 'folder',
            id: '1',
            payload: {},
            children: [{ type: 'folder', id: '2' }],
        }).issues).toBeUndefined();
    });
    (0, vitest_1.describe)('assertResourceType', () => {
        (0, vitest_1.test)('should not throw for valid resource type', () => {
            const validResource = {
                id: 'test-id',
                type: 'example-request',
                payload: { url: 'https://example.com' },
                children: [
                    {
                        type: 'example-message',
                        payload: { content: 'Hello' },
                    },
                ],
            };
            (0, vitest_1.expect)(() => (0, item_type_js_1.assertResourceType)(ExampleRequest, validResource)).not.toThrow();
        });
        (0, vitest_1.test)('should not throw for invalid resource type', () => {
            const invalidResource = {
                type: 'example-request',
                payload: { url: 'https://example.com' },
                children: [
                    {
                        type: 'example-message',
                        payload: { content: 'Hello' },
                    },
                ],
            };
            (0, vitest_1.expect)(() => (0, item_type_js_1.assertResourceType)(ExampleRequest, invalidResource)).toThrow('Invalid input for model');
        });
    });
});
