// @postman/runtime/dist/items/collection/collection-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.collection = void 0;
const runtime_models_1 = require("@postman/runtime.models");
exports.collection = {
    name: 'collection',
    items: {
        collection(item, run) {
            (0, runtime_models_1.assertResourceType)(runtime_models_1.Collection, item);
            run.enqueue({
                type: 'collection:execute',
            });
        },
    },
    execute: {
        'collection:execute'(_event, run) {
            (0, runtime_models_1.assertResourceType)(runtime_models_1.Collection, run.item);
            const collection = {
                ...run.item,
                id: run.item.id ?? crypto.randomUUID(),
            };
            if (!collection.children)
                return;
            run.waitUntil(async () => {
                const plan = planCollectionRun(collection);
                for await (const { folders, item } of plan) {
                    const subprocess = run.execute(item, {
                        collection,
                        folders,
                    });
                    await subprocess.complete;
                }
            });
        },
    },
};
/**
 * Flatten out collection and folders into run
 *
 * TODO
 * - "profile" for reusable collection run
 * - `skipRequest` support
 */
function planCollectionRun(collection) {
    if (!collection.children)
        return [];
    // 1. Recursively flatten array, building up chain of folders
    const extractChildren = (folders, folder) => {
        if (!folder.children)
            return [];
        const folderWithId = {
            ...folder,
            id: folder.id ?? crypto.randomUUID(),
        };
        return folder.children?.flatMap((child) => child.type === 'folder' ?
            extractChildren([...folders, folderWithId], child)
            : [{ folders: [...folders, folderWithId], item: child }]);
    };
    const plan = collection.children.flatMap((child) => child.type === 'folder' ?
        extractChildren([], child)
        : [{ folders: [], item: child }]);
    return plan;
}
